
<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Lecture 10 - Property &amp; Reaction Prediction &#8212; My sample book</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.4045f2051d55cab465a707391d5b2007.min.css?v=6ad1a40c" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=2b91c5f0"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js?v=afe5de03"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'lecture-10-old';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.png" class="logo__image only-light" alt="My sample book - Home"/>
    <script>document.write(`<img src="_static/logo.png" class="logo__image only-dark" alt="My sample book - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Welcome to CHEM 5080: AI for Chemistry
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="lecture-01.html">Lecture 1 - Python Primer</a></li>
<li class="toctree-l1"><a class="reference internal" href="lecture-02.html">Lecture 2 - Pandas and Plotting</a></li>
<li class="toctree-l1"><a class="reference internal" href="lecture-03.html">Lecture 3 - SMILES and RDKit</a></li>
<li class="toctree-l1"><a class="reference internal" href="lecture-04.html">Lecture 4 - Chemical Structure Identifier</a></li>
<li class="toctree-l1"><a class="reference internal" href="lecture-05.html">Lecture 5 - Regression and Classification</a></li>
<li class="toctree-l1"><a class="reference internal" href="lecture-06.html">Lecture 6 - Cross-Validation</a></li>
<li class="toctree-l1"><a class="reference internal" href="lecture-07.html">Lecture 7 - Decision Trees and Random Forests</a></li>
<li class="toctree-l1"><a class="reference internal" href="lecture-08.html">Lecture 8 - Neural Networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="lecture-09.html">Lecture 9 - Graph Neural Networks</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2Flecture-10-old.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/lecture-10-old.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Lecture 10 - Property & Reaction Prediction</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-goals">Learning goals</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#directed-message-passing-neural-network-d-mpnn">1. Directed message-passing neural network (D-MPNN)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-preparation">2. Data preparation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#property-prediction-regression">3. Property prediction (regression)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pick-blocks">3.1 Pick blocks</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#build-model-and-trainer">3.2 Build model and trainer</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#train">3.3 Train</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#test-and-parity-plot">3.4 Test and parity plot</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#property-prediction-classification">4. Property prediction (classification)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#build-classification-dataset">4.1 Build classification dataset</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#model-and-training">4.2 Model and training</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#roc-curve">4.3 ROC curve</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reactivity-and-selectivity">5. Reactivity and selectivity</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reactivity-classifier">5.1 Reactivity classifier</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#atom-level-selectivity-build-labels-per-atom">5.2 Atom-level selectivity: build labels per atom</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#rationale">6. Rationale</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#solubility-prediction-rationale">6.1 Solubility prediction rationale</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reactivity-prediction-rationale">6.2 Reactivity prediction rationale</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#chemprop-cli-command-line-interface">7. Chemprop CLI (Command-Line Interface)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reactivity-classification-ch-oxidation-dataset">Reactivity classification (C–H oxidation dataset)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#glossary-references-in-class-activity-with-solutions">10. Glossary • References • In-class activity with solutions</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#glossary">10.1 Glossary</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#references">10.2 References</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#in-class-activity">10.3 In-class activity</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#solutions">10.4 Solutions</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="lecture-10-property-reaction-prediction">
<h1>Lecture 10 - Property &amp; Reaction Prediction<a class="headerlink" href="#lecture-10-property-reaction-prediction" title="Permalink to this heading">#</a></h1>
<nav class="contents local" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#learning-goals" id="id1">Learning goals</a></p></li>
<li><p><a class="reference internal" href="#directed-message-passing-neural-network-d-mpnn" id="id2">1. Directed message-passing neural network (D-MPNN)</a></p></li>
<li><p><a class="reference internal" href="#data-preparation" id="id3">2. Data preparation</a></p></li>
<li><p><a class="reference internal" href="#property-prediction-regression" id="id4">3. Property prediction (regression)</a></p></li>
<li><p><a class="reference internal" href="#property-prediction-classification" id="id5">4. Property prediction (classification)</a></p></li>
<li><p><a class="reference internal" href="#reactivity-and-selectivity" id="id6">5. Reactivity and selectivity</a></p></li>
<li><p><a class="reference internal" href="#rationale" id="id7">6. Rationale</a></p></li>
<li><p><a class="reference internal" href="#chemprop-cli-command-line-interface" id="id8">7. Chemprop CLI (Command-Line Interface)</a></p></li>
<li><p><a class="reference internal" href="#glossary-references-in-class-activity-with-solutions" id="id9">10. Glossary • References • In-class activity with solutions</a></p></li>
</ul>
</nav>
<section id="learning-goals">
<h2><a class="toc-backref" href="#id1" role="doc-backlink">Learning goals</a><a class="headerlink" href="#learning-goals" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>Set up <strong>Chemprop v2</strong> for regression and classification on our C-H oxidation dataset.</p></li>
<li><p>Train four <strong>single task</strong> models for: Solubility, pKa, Melting Point, Toxicity.</p></li>
<li><p>Train a <strong>reactivity</strong> classifier and an <strong>atom-level selectivity</strong> predictor.</p></li>
<li><p>Interpret a trained model with <strong>Shapley values (SHAP)</strong> at the feature and node levels.</p></li>
</ul>
<div class="admonition-important-note admonition">
<p class="admonition-title">Important note</p>
<p>For this lecture 10, it is recommended to run everything in Colab. On this HTML page, some outputs are disabled due to execution limits, so only code and some example output is displayed.</p>
</div>
<hr class="docutils" />
<div class="cell tag_hide-input docutils container">
<details class="hide above-input">
<summary aria-label="Toggle hidden content">
<span class="collapsed">Show code cell source</span>
<span class="expanded">Hide code cell source</span>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1. Setup</span>

<span class="c1"># Hide code cell source</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span><span class="o">,</span><span class="w"> </span><span class="nn">os</span><span class="o">,</span><span class="w"> </span><span class="nn">sys</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>

<span class="c1"># RDKit (must be installed in THIS venv)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">rdkit</span><span class="w"> </span><span class="kn">import</span> <span class="n">Chem</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">rdkit.Chem</span><span class="w"> </span><span class="kn">import</span> <span class="n">Draw</span><span class="p">,</span> <span class="n">Descriptors</span><span class="p">,</span> <span class="n">Crippen</span><span class="p">,</span> <span class="n">rdMolDescriptors</span><span class="p">,</span> <span class="n">AllChem</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span><span class="o">,</span><span class="w"> </span><span class="nn">math</span><span class="o">,</span><span class="w"> </span><span class="nn">os</span><span class="o">,</span><span class="w"> </span><span class="nn">sys</span><span class="o">,</span><span class="w"> </span><span class="nn">json</span><span class="o">,</span><span class="w"> </span><span class="nn">time</span><span class="o">,</span><span class="w"> </span><span class="nn">random</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">networkx</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">nx</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>  
<span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>



<span class="c1"># Sklearn bits for splitting and metrics</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.model_selection</span><span class="w"> </span><span class="kn">import</span> <span class="n">train_test_split</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.preprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">StandardScaler</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.pipeline</span><span class="w"> </span><span class="kn">import</span> <span class="n">Pipeline</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.compose</span><span class="w"> </span><span class="kn">import</span> <span class="n">TransformedTargetRegressor</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.tree</span><span class="w"> </span><span class="kn">import</span> <span class="n">DecisionTreeRegressor</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.ensemble</span><span class="w"> </span><span class="kn">import</span> <span class="n">RandomForestRegressor</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.neural_network</span><span class="w"> </span><span class="kn">import</span> <span class="n">MLPRegressor</span>


<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span><span class="n">mean_squared_error</span><span class="p">,</span> <span class="n">mean_absolute_error</span><span class="p">,</span> <span class="n">r2_score</span><span class="p">,</span>
                             <span class="n">accuracy_score</span><span class="p">,</span> <span class="n">precision_score</span><span class="p">,</span> <span class="n">recall_score</span><span class="p">,</span>
                             <span class="n">f1_score</span><span class="p">,</span> <span class="n">roc_auc_score</span><span class="p">,</span> <span class="n">roc_curve</span><span class="p">,</span> <span class="n">confusion_matrix</span><span class="p">)</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;chemprop.data.splitting&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">ERROR</span><span class="p">)</span>

<span class="c1"># RDKit</span>
<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">rdkit</span><span class="w"> </span><span class="kn">import</span> <span class="n">Chem</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">rdkit.Chem</span><span class="w"> </span><span class="kn">import</span> <span class="n">Draw</span><span class="p">,</span> <span class="n">Descriptors</span><span class="p">,</span> <span class="n">Crippen</span><span class="p">,</span> <span class="n">rdMolDescriptors</span><span class="p">,</span> <span class="n">AllChem</span>
<span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="o">%</span><span class="k">pip</span> install rdkit
        <span class="kn">from</span><span class="w"> </span><span class="nn">rdkit</span><span class="w"> </span><span class="kn">import</span> <span class="n">Chem</span>
        <span class="kn">from</span><span class="w"> </span><span class="nn">rdkit.Chem</span><span class="w"> </span><span class="kn">import</span> <span class="n">Draw</span><span class="p">,</span> <span class="n">Descriptors</span><span class="p">,</span> <span class="n">Crippen</span><span class="p">,</span> <span class="n">rdMolDescriptors</span><span class="p">,</span> <span class="n">AllChem</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;RDKit is not available in this environment. Drawing and descriptors will be skipped.&quot;</span><span class="p">)</span>
        <span class="n">Chem</span> <span class="o">=</span> <span class="kc">None</span>
</pre></div>
</div>
</div>
</details>
</div>
</section>
<section id="directed-message-passing-neural-network-d-mpnn">
<h2><a class="toc-backref" href="#id2" role="doc-backlink">1. Directed message-passing neural network (D-MPNN)</a><a class="headerlink" href="#directed-message-passing-neural-network-d-mpnn" title="Permalink to this heading">#</a></h2>
<p>We will train models for four molecular properties and reaction-related labels using <strong>Chemprop</strong>.</p>
<p>Briefly speaking, Chemprop builds neural models for molecules using a <em>directed</em> message passing neural network (D-MPNN).</p>
<p>As you recall from previous lecture, a message passing neural network (MPNN) updates hidden vectors on nodes and edges with local neighbor information, then an aggregation step creates a graph-level vector for prediction.</p>
<p>Chemprop’s directed variant changes the way messages flow: instead of passing information back and forth between atoms, it assigns a hidden state to each directed bond (atom <code class="docutils literal notranslate"><span class="pre">i</span></code> → atom <code class="docutils literal notranslate"><span class="pre">j</span></code>). This prevents immediate backtracking (“tottering”) where messages would simply bounce between two atoms without capturing new context. By using directed bonds, the model distinguishes subtle chemical environments. For example, the information carried from a carbon toward a nitrogen can be different than the reverse direction, which matters for reactivity and selectivity.</p>
<p>As a GNN, Chemprop also featurizes a molecule as a <strong>graph</strong>:</p>
<ul class="simple">
<li><p><strong>Nodes</strong> are atoms with features like atomic number, degree, aromaticity.</p></li>
<li><p><strong>Edges</strong> are bonds with features like bond order and stereo.</p></li>
</ul>
<p>Initial directed bond state <span class="math notranslate nohighlight">\(h_{i→j}^{(0)}\)</span> is a learned function of the source atom features and the bond features. For <code class="docutils literal notranslate"><span class="pre">t</span> <span class="pre">=</span> <span class="pre">1..T</span></code>, update
<span class="math notranslate nohighlight">\(
h_{i \to j}^{(t)} = \sigma \Big( W \cdot \big( h_{i \to j}^{(t-1)} + \sum_{k \in \mathcal{N}(i) \setminus \{j\}} h_{k \to i}^{(t-1)} \big) + b \Big)
\)</span>,</p>
<p>where <code class="docutils literal notranslate"><span class="pre">σ</span></code> is an activation such as <code class="docutils literal notranslate"><span class="pre">ReLU</span></code>, <code class="docutils literal notranslate"><span class="pre">W</span></code> is a learned weight, <span class="math notranslate nohighlight">\(x_{i→j}\)</span> are featurized inputs, <span class="math notranslate nohighlight">\(⊕\)</span> is concatenation, and <span class="math notranslate nohighlight">\(N(i)\{j}\)</span> removes the target atom to avoid immediate backtracking. After T steps, Chemprop aggregates per directed bond states to atom states, then pools to a molecule vector <span class="math notranslate nohighlight">\(h_mol\)</span> using sum or mean or attention pooling. <span class="math notranslate nohighlight">\(h_mol\)</span> feeds a multitask feedforward head.</p>
<p>We have been working with the following quite many times:</p>
<ul class="simple">
<li><p><strong>Solubility_mol_per_L</strong>: continuous. Regression with loss like MSE or MAE.</p></li>
<li><p><strong>pKa</strong>: continuous. Regression.</p></li>
<li><p><strong>Melting Point</strong>: continuous. Regression.</p></li>
<li><p><strong>Toxicity</strong>: categorical with values like <code class="docutils literal notranslate"><span class="pre">toxic</span></code> or <code class="docutils literal notranslate"><span class="pre">non_toxic</span></code>. Binary classification.</p></li>
</ul>
<p>While these two we never try before:</p>
<ul class="simple">
<li><p><strong>Reactivity</strong>: binary label <code class="docutils literal notranslate"><span class="pre">1</span></code> vs <code class="docutils literal notranslate"><span class="pre">-1</span></code>. Binary classification. In our C-H oxidation dataset, this means whether the substrate will undergo oxidation.</p></li>
<li><p><strong>Site Selectivity</strong>: a set of atom indices. Atom-level classification inside a molecule. In our C-H oxidation dataset, this means which atom(s) are most likely to oxidize under certain electrochemical reaction condition, expressed as atom indices in the SMILES.</p></li>
</ul>
<p>As a reminder, below are some reference formulas:</p>
<ul class="simple">
<li><p>Regression losses<br />
$<span class="math notranslate nohighlight">\(
\text{MSE} = \frac{1}{n}\sum_i (y_i - \hat y_i)^2,\qquad
\text{MAE} = \frac{1}{n}\sum_i |y_i - \hat y_i|
\)</span>$</p></li>
<li><p>Binary cross entropy<br />
$<span class="math notranslate nohighlight">\(
\mathcal{L} = -\frac{1}{n}\sum_i \big(y_i\log \hat p_i + (1-y_i)\log(1-\hat p_i)\big)
\)</span>$</p></li>
</ul>
<blockquote>
<div><p>You saw this idea in earlier lectures. The new part is that Chemprop builds the graph from SMILES and offers modules for molecule, reaction and atom/bond tasks.</p>
</div></blockquote>
</section>
<hr class="docutils" />
<section id="data-preparation">
<h2><a class="toc-backref" href="#id3" role="doc-backlink">2. Data preparation</a><a class="headerlink" href="#data-preparation" title="Permalink to this heading">#</a></h2>
<p>We begin with load and inspect the C-H oxidation dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;https://raw.githubusercontent.com/zzhenglab/ai4chem/main/book/_data/C_H_oxidation_dataset.csv&quot;</span>
<span class="n">df_raw</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
<span class="n">df_raw</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Compound Name</th>
      <th>CAS</th>
      <th>SMILES</th>
      <th>Solubility_mol_per_L</th>
      <th>pKa</th>
      <th>Toxicity</th>
      <th>Melting Point</th>
      <th>Reactivity</th>
      <th>Oxidation Site</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>3,4-dihydro-1H-isochromene</td>
      <td>493-05-0</td>
      <td>c1ccc2c(c1)CCOC2</td>
      <td>0.103906</td>
      <td>5.80</td>
      <td>non_toxic</td>
      <td>65.8</td>
      <td>1</td>
      <td>8,10</td>
    </tr>
    <tr>
      <th>1</th>
      <td>9H-fluorene</td>
      <td>86-73-7</td>
      <td>c1ccc2c(c1)Cc1ccccc1-2</td>
      <td>0.010460</td>
      <td>5.82</td>
      <td>toxic</td>
      <td>90.0</td>
      <td>1</td>
      <td>7</td>
    </tr>
    <tr>
      <th>2</th>
      <td>1,2,3,4-tetrahydronaphthalene</td>
      <td>119-64-2</td>
      <td>c1ccc2c(c1)CCCC2</td>
      <td>0.020589</td>
      <td>5.74</td>
      <td>toxic</td>
      <td>69.4</td>
      <td>1</td>
      <td>7,10</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ethylbenzene</td>
      <td>100-41-4</td>
      <td>CCc1ccccc1</td>
      <td>0.048107</td>
      <td>5.87</td>
      <td>non_toxic</td>
      <td>65.0</td>
      <td>1</td>
      <td>1,2</td>
    </tr>
    <tr>
      <th>4</th>
      <td>cyclohexene</td>
      <td>110-83-8</td>
      <td>C1=CCCCC1</td>
      <td>0.060688</td>
      <td>5.66</td>
      <td>non_toxic</td>
      <td>96.4</td>
      <td>1</td>
      <td>3,6</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Basic info</span>
<span class="n">df_raw</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">df_raw</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>((575, 9),
 [&#39;Compound Name&#39;,
  &#39;CAS&#39;,
  &#39;SMILES&#39;,
  &#39;Solubility_mol_per_L&#39;,
  &#39;pKa&#39;,
  &#39;Toxicity&#39;,
  &#39;Melting Point&#39;,
  &#39;Reactivity&#39;,
  &#39;Oxidation Site&#39;])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Clean a copy and normalize a few columns</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df_raw</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="c1"># Toxicity -&gt; binary string &#39;toxic&#39;/&#39;non_toxic&#39; to 1/0 if present</span>
<span class="n">tox_map</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;toxic&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;non_toxic&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">}</span>
<span class="k">if</span> <span class="s2">&quot;Toxicity&quot;</span> <span class="ow">in</span> <span class="n">df</span><span class="p">:</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tox_bin&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;Toxicity&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">tox_map</span><span class="p">)</span>

<span class="c1"># Reactivity -&gt; 1/-1 to 1/0</span>
<span class="k">if</span> <span class="s2">&quot;Reactivity&quot;</span> <span class="ow">in</span> <span class="n">df</span><span class="p">:</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;react_bin&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;Reactivity&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">1</span> <span class="k">if</span> <span class="n">x</span><span class="o">==</span><span class="mi">1</span> <span class="k">else</span> <span class="mi">0</span><span class="p">)</span>

<span class="c1"># Oxidation Site -&gt; list of ints</span>
<span class="k">def</span><span class="w"> </span><span class="nf">parse_sites</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)]</span>
    <span class="k">return</span> <span class="p">[]</span>
<span class="n">df</span><span class="p">[</span><span class="s2">&quot;site_list&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;Oxidation Site&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">parse_sites</span><span class="p">)</span>

<span class="c1"># Take log of solubility (keep same column name)</span>
<span class="k">if</span> <span class="s2">&quot;Solubility_mol_per_L&quot;</span> <span class="ow">in</span> <span class="n">df</span><span class="p">:</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;logS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;Solubility_mol_per_L&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="mf">1e-6</span><span class="p">)</span>

<span class="n">df</span><span class="p">[[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">,</span><span class="s2">&quot;logS&quot;</span><span class="p">,</span><span class="s2">&quot;pKa&quot;</span><span class="p">,</span><span class="s2">&quot;Toxicity&quot;</span><span class="p">,</span><span class="s2">&quot;Melting Point&quot;</span><span class="p">,</span><span class="s2">&quot;react_bin&quot;</span><span class="p">,</span><span class="s2">&quot;site_list&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>SMILES</th>
      <th>logS</th>
      <th>pKa</th>
      <th>Toxicity</th>
      <th>Melting Point</th>
      <th>react_bin</th>
      <th>site_list</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>c1ccc2c(c1)CCOC2</td>
      <td>-0.983356</td>
      <td>5.80</td>
      <td>non_toxic</td>
      <td>65.8</td>
      <td>1</td>
      <td>[8, 10]</td>
    </tr>
    <tr>
      <th>1</th>
      <td>c1ccc2c(c1)Cc1ccccc1-2</td>
      <td>-1.980414</td>
      <td>5.82</td>
      <td>toxic</td>
      <td>90.0</td>
      <td>1</td>
      <td>[7]</td>
    </tr>
    <tr>
      <th>2</th>
      <td>c1ccc2c(c1)CCCC2</td>
      <td>-1.686343</td>
      <td>5.74</td>
      <td>toxic</td>
      <td>69.4</td>
      <td>1</td>
      <td>[7, 10]</td>
    </tr>
    <tr>
      <th>3</th>
      <td>CCc1ccccc1</td>
      <td>-1.317782</td>
      <td>5.87</td>
      <td>non_toxic</td>
      <td>65.0</td>
      <td>1</td>
      <td>[1, 2]</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C1=CCCCC1</td>
      <td>-1.216890</td>
      <td>5.66</td>
      <td>non_toxic</td>
      <td>96.4</td>
      <td>1</td>
      <td>[3, 6]</td>
    </tr>
    <tr>
      <th>5</th>
      <td>C1CCSC1</td>
      <td>-0.917634</td>
      <td>5.97</td>
      <td>non_toxic</td>
      <td>15.8</td>
      <td>1</td>
      <td>[3, 5]</td>
    </tr>
    <tr>
      <th>6</th>
      <td>CN1CCCC1=O</td>
      <td>-0.499442</td>
      <td>5.91</td>
      <td>non_toxic</td>
      <td>71.1</td>
      <td>1</td>
      <td>[3]</td>
    </tr>
    <tr>
      <th>7</th>
      <td>COCc1ccccc1</td>
      <td>-1.070756</td>
      <td>5.61</td>
      <td>non_toxic</td>
      <td>108.5</td>
      <td>1</td>
      <td>[3]</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="admonition-exercise-1 admonition">
<p class="admonition-title">⏰ <strong>Exercise 1</strong></p>
<p>Count number of postive and negative reaction outcomes in <code class="docutils literal notranslate"><span class="pre">react_bin</span></code>..</p>
</div>
<p>We will create <strong>MoleculeDatapoint</strong> objects from SMILES and targets, split the data, and build loaders.
Our first target will be <code class="docutils literal notranslate"><span class="pre">solubility</span></code>.</p>
<blockquote>
<div><p>Step 1. Build datapoints.</p>
</div></blockquote>
<p>Each row of the dataframe is now represented as a <code class="docutils literal notranslate"><span class="pre">MoleculeDatapoint</span></code>. It stores the SMILES, the numeric target (solubility here), plus metadata like optional weights.<br />
This is the <em>atomic unit</em> Chemprop will pass to the featurizer.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Keep rows that have both SMILES and solubility</span>
<span class="n">df_sol</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">,</span><span class="s2">&quot;logS&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
<span class="n">smis</span> <span class="o">=</span> <span class="n">df_sol</span><span class="p">[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">ys</span>   <span class="o">=</span> <span class="n">df_sol</span><span class="p">[</span><span class="s2">&quot;logS&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>

<span class="n">sol_datapoints</span> <span class="o">=</span> <span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">MoleculeDatapoint</span><span class="o">.</span><span class="n">from_smi</span><span class="p">(</span><span class="n">smi</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span> <span class="k">for</span> <span class="n">smi</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">smis</span><span class="p">,</span> <span class="n">ys</span><span class="p">)]</span>
<span class="nb">len</span><span class="p">(</span><span class="n">sol_datapoints</span><span class="p">),</span> <span class="n">sol_datapoints</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">y</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
<p>Step 2. Split to train, val, test.</p>
<p>We divided the list of datapoints into three folds.</p>
<ul class="simple">
<li><p>Training: used to fit model weights.</p></li>
<li><p>Validation: used to monitor progress and stop early.</p></li>
<li><p>Test: kept blind until the end.<br />
Even though we used a random split here, Chemprop also supports scaffold-based splits which are often better for chemistry.</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="n">dp</span><span class="o">.</span><span class="n">mol</span> <span class="k">for</span> <span class="n">dp</span> <span class="ow">in</span> <span class="n">sol_datapoints</span><span class="p">]</span>

<span class="n">train_lists</span><span class="p">,</span> <span class="n">val_lists</span><span class="p">,</span> <span class="n">test_lists</span> <span class="o">=</span> <span class="n">make_split_indices</span><span class="p">(</span>
    <span class="n">mols</span><span class="o">=</span><span class="n">mols</span><span class="p">,</span>
    <span class="n">split</span><span class="o">=</span><span class="s2">&quot;random&quot;</span><span class="p">,</span>  <span class="c1"># or &quot;scaffold&quot;, &quot;stratified&quot;, etc.</span>
    <span class="n">sizes</span><span class="o">=</span><span class="p">(</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span>
    <span class="n">seed</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">num_replicates</span><span class="o">=</span><span class="mi">1</span>
<span class="p">)</span>

<span class="n">train_dpss</span><span class="p">,</span> <span class="n">val_dpss</span><span class="p">,</span> <span class="n">test_dpss</span> <span class="o">=</span> <span class="n">split_data_by_indices</span><span class="p">(</span>
    <span class="n">sol_datapoints</span><span class="p">,</span> <span class="n">train_lists</span><span class="p">,</span> <span class="n">val_lists</span><span class="p">,</span> <span class="n">test_lists</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">train_dpss</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="nb">len</span><span class="p">(</span><span class="n">val_dpss</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="nb">len</span><span class="p">(</span><span class="n">test_dpss</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
</pre></div>
</div>
<blockquote>
<div><p>Step 3. Build dataset objects and scale targets.
A <code class="docutils literal notranslate"><span class="pre">MoleculeDataset</span></code> wraps the datapoints and applies the chosen featurizer.</p>
</div></blockquote>
<ul class="simple">
<li><p>Here we used <code class="docutils literal notranslate"><span class="pre">SimpleMoleculeMolGraphFeaturizer</span></code>, which turns atoms and bonds into numeric arrays.</p></li>
<li><p>We also normalized the target values (subtract mean, divide by std) so the model trains smoothly. The stored <code class="docutils literal notranslate"><span class="pre">scaler</span></code> allows us to unscale predictions back.</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">feat</span> <span class="o">=</span> <span class="n">featurizers</span><span class="o">.</span><span class="n">SimpleMoleculeMolGraphFeaturizer</span><span class="p">()</span>
<span class="n">train_set</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MoleculeDataset</span><span class="p">(</span><span class="n">train_dpss</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>
<span class="n">scaler</span> <span class="o">=</span> <span class="n">train_set</span><span class="o">.</span><span class="n">normalize_targets</span><span class="p">()</span>  <span class="c1"># store mean/var</span>

<span class="n">val_set</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MoleculeDataset</span><span class="p">(</span><span class="n">val_dpss</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>
<span class="n">val_set</span><span class="o">.</span><span class="n">normalize_targets</span><span class="p">(</span><span class="n">scaler</span><span class="p">)</span>

<span class="n">test_set</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MoleculeDataset</span><span class="p">(</span><span class="n">test_dpss</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>

<span class="c1"># Peek at one item structure</span>
<span class="n">item0</span> <span class="o">=</span> <span class="n">train_set</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="nb">type</span><span class="p">(</span><span class="n">item0</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span> <span class="n">item0</span><span class="o">.</span><span class="n">y</span><span class="p">,</span> <span class="n">item0</span><span class="o">.</span><span class="n">mg</span><span class="o">.</span><span class="n">V</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">item0</span><span class="o">.</span><span class="n">mg</span><span class="o">.</span><span class="n">E</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
<blockquote>
<div><p>Step 4. Dataloaders.
Finally, we wrapped datasets in PyTorch-style <code class="docutils literal notranslate"><span class="pre">DataLoader</span></code>s.</p>
</div></blockquote>
<ul class="simple">
<li><p>Training loader will shuffle each epoch.</p></li>
<li><p>Validation and test loaders do not shuffle, to keep evaluation consistent.<br />
Batching is automatic: molecules of different sizes are packed together and masks are used internally.</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">train_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">train_set</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">val_loader</span>   <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">val_set</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">test_loader</span>  <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">test_set</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition-tip admonition">
<p class="admonition-title">Tip</p>
<p>Scaling targets helps training for many regressors. We keep the unscale transform for outputs inside the model so that you get predictions in the original units.</p>
</div>
</section>
<hr class="docutils" />
<section id="property-prediction-regression">
<h2><a class="toc-backref" href="#id4" role="doc-backlink">3. Property prediction (regression)</a><a class="headerlink" href="#property-prediction-regression" title="Permalink to this heading">#</a></h2>
<p>We will configure a small MPNN for regression.</p>
<p>In particular, we will:</p>
<ol class="arabic simple">
<li><p>Choose the neural blocks that define how messages are passed, pooled, and transformed into outputs.</p></li>
<li><p>Assemble them into a complete model object.</p></li>
<li><p>Set up a training loop with early stopping and checkpoints.</p></li>
<li><p>Evaluate predictions on a held-out test set and visualize the quality using a parity plot.</p></li>
</ol>
<section id="pick-blocks">
<h3>3.1 Pick blocks<a class="headerlink" href="#pick-blocks" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">mp</span>  <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">BondMessagePassing</span><span class="p">()</span>        <span class="c1"># node/edge update</span>
<span class="n">agg</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">MeanAggregation</span><span class="p">()</span>           <span class="c1"># pool node vectors</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">RegressionFFN</span><span class="p">(</span>              <span class="c1"># simple FFN head</span>
    <span class="n">output_transform</span><span class="o">=</span><span class="n">nn</span><span class="o">.</span><span class="n">UnscaleTransform</span><span class="o">.</span><span class="n">from_standard_scaler</span><span class="p">(</span><span class="n">scaler</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">batch_norm</span> <span class="o">=</span> <span class="kc">True</span>
<span class="n">metrics</span> <span class="o">=</span> <span class="p">[</span><span class="n">nn</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">RMSE</span><span class="p">(),</span> <span class="n">nn</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">MAE</span><span class="p">()]</span>  <span class="c1"># first metric used for early stopping</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong>BondMessagePassing()</strong> updates hidden states on each directed bond by passing information across neighbors.</p></li>
<li><p><strong>MeanAggregation()</strong> pools hidden vectors to form atom or molecule-level representations. Other options like sum or attention pooling are possible.</p></li>
<li><p><strong>RegressionFFN()</strong> is a feed-forward head. Here we attach an <code class="docutils literal notranslate"><span class="pre">UnscaleTransform</span></code> so predictions can be mapped back to the original solubility scale.</p></li>
<li><p><strong>Batch normalization</strong> improves stability by normalizing hidden states during training.</p></li>
<li><p><strong>Metrics</strong> let us monitor training. RMSE (root mean squared error) and MAE (mean absolute error) are both useful, but RMSE is often more sensitive to large errors and is used for early stopping.</p></li>
</ul>
</section>
<section id="build-model-and-trainer">
<h3>3.2 Build model and trainer<a class="headerlink" href="#build-model-and-trainer" title="Permalink to this heading">#</a></h3>
<p>Once the blocks are chosen, we wrap them into a full MPNN model. Chemprop uses PyTorch Lightning under the hood, so we also set up a Trainer:</p>
<ul class="simple">
<li><p>The ModelCheckpoint callback saves the best version of the model during training, based on validation loss.</p></li>
<li><p>The trainer can run on CPU or GPU (<code class="docutils literal notranslate"><span class="pre">accelerator=&quot;auto&quot;</span></code>).</p></li>
</ul>
<p>We limit to 15 epochs here for demonstration, but in practice you might extend this depending on dataset size and convergence.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">mpnn_sol</span> <span class="o">=</span> <span class="n">models</span><span class="o">.</span><span class="n">MPNN</span><span class="p">(</span><span class="n">mp</span><span class="p">,</span> <span class="n">agg</span><span class="p">,</span> <span class="n">out</span><span class="p">,</span> <span class="n">batch_norm</span><span class="p">,</span> <span class="n">metrics</span><span class="p">)</span>

<span class="n">checkpoint_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;checkpoints_sol&quot;</span><span class="p">)</span>
<span class="n">checkpoint_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">ckpt</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">callbacks</span><span class="o">.</span><span class="n">ModelCheckpoint</span><span class="p">(</span>
    <span class="n">dirpath</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">checkpoint_dir</span><span class="p">),</span> <span class="n">filename</span><span class="o">=</span><span class="s2">&quot;best-</span><span class="si">{epoch}</span><span class="s2">-</span><span class="si">{val_loss:.3f}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="n">monitor</span><span class="o">=</span><span class="s2">&quot;val_loss&quot;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;min&quot;</span><span class="p">,</span> <span class="n">save_last</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="n">trainer</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">Trainer</span><span class="p">(</span>
    <span class="n">logger</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">enable_checkpointing</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">accelerator</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span>
    <span class="n">devices</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_epochs</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">callbacks</span><span class="o">=</span><span class="p">[</span><span class="n">ckpt</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">mpnn_sol</span>
</pre></div>
</div>
<p>At this stage, we have a complete pipeline: dataset loaders, model blocks, and a trainer that knows when to save progress.</p>
</section>
<section id="train">
<h3>3.3 Train<a class="headerlink" href="#train" title="Permalink to this heading">#</a></h3>
<p>Since we implement everything earlier, now training is as simple as calling <code class="docutils literal notranslate"><span class="pre">fit()</span></code>. The trainer will:</p>
<ol class="arabic simple">
<li><p>Iterate over the training loader each epoch.</p></li>
<li><p>Evaluate on the validation loader.</p></li>
<li><p>Save checkpoints when the validation RMSE improves.</p></li>
</ol>
<p>During training, you can monitor validation loss to see whether the model is underfitting, overfitting, or converging as expected.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">trainer</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">mpnn_sol</span><span class="p">,</span> <span class="n">train_loader</span><span class="p">,</span> <span class="n">val_loader</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="test-and-parity-plot">
<h3>3.4 Test and parity plot<a class="headerlink" href="#test-and-parity-plot" title="Permalink to this heading">#</a></h3>
<p>After training, we hold back the test set for final evaluation. We then visualize predicted vs. true values with a parity plot:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">results</span> <span class="o">=</span> <span class="n">trainer</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">mpnn_sol</span><span class="p">,</span> <span class="n">test_loader</span><span class="p">)</span>

<span class="c1"># Gather predictions for parity</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>
<span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">inference_mode</span><span class="p">():</span>
    <span class="n">preds</span> <span class="o">=</span> <span class="n">trainer</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">mpnn_sol</span><span class="p">,</span> <span class="n">test_loader</span><span class="p">)</span>
<span class="n">preds</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">preds</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>

<span class="n">y_true</span> <span class="o">=</span> <span class="n">test_set</span><span class="o">.</span><span class="n">Y</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Test size:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">y_true</span><span class="p">))</span>

<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">preds</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>

<span class="c1"># Set both axes to the same range</span>
<span class="n">lims</span> <span class="o">=</span> <span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">lims</span><span class="p">,</span> <span class="n">lims</span><span class="p">,</span> <span class="s2">&quot;k--&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="n">lims</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="n">lims</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;True solubility (mol/L)&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Predicted&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Parity plot: Solubility&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<div class="admonition-output admonition">
<p class="admonition-title">Output</p>
<p>│         test/mae          │    0.2726641297340393     │
│         test/rmse         │    0.4700523018836975     │</p>
<p>Test size: 58</p>
</div>
<div class="cell tag_hide-input docutils container">
<details class="hide above-input">
<summary aria-label="Toggle hidden content">
<span class="collapsed">Show code cell source</span>
<span class="expanded">Hide code cell source</span>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">IPython.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">Image</span><span class="p">,</span> <span class="n">display</span>
<span class="n">display</span><span class="p">(</span><span class="n">Image</span><span class="p">(</span><span class="n">url</span><span class="o">=</span><span class="s2">&quot;https://raw.githubusercontent.com/zzhenglab/ai4chem/main/book/_data/lec-10-pa-plot.png&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<div class="output text_html"><img src="https://raw.githubusercontent.com/zzhenglab/ai4chem/main/book/_data/lec-10-pa-plot.png"/></div></div>
</div>
<div class="admonition-exercise admonition">
<p class="admonition-title">⏰ Exercise</p>
<p>Change the aggregation from <code class="docutils literal notranslate"><span class="pre">MeanAggregation()</span></code> to <code class="docutils literal notranslate"><span class="pre">SumAggregation()</span></code> and retrain for 10 epochs. Compare RMSE and the parity plot. What changed?</p>
</div>
</section>
</section>
<hr class="docutils" />
<section id="property-prediction-classification">
<h2><a class="toc-backref" href="#id5" role="doc-backlink">4. Property prediction (classification)</a><a class="headerlink" href="#property-prediction-classification" title="Permalink to this heading">#</a></h2>
<p>Here we predict <code class="docutils literal notranslate"><span class="pre">toxic</span></code> vs <code class="docutils literal notranslate"><span class="pre">non_toxic</span></code>.</p>
<section id="build-classification-dataset">
<h3>4.1 Build classification dataset<a class="headerlink" href="#build-classification-dataset" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">df_tox</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">,</span><span class="s2">&quot;tox_bin&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
<span class="n">smis</span> <span class="o">=</span> <span class="n">df_tox</span><span class="p">[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">ys</span>   <span class="o">=</span> <span class="n">df_tox</span><span class="p">[</span><span class="s2">&quot;tox_bin&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>

<span class="n">tox_dps</span> <span class="o">=</span> <span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">MoleculeDatapoint</span><span class="o">.</span><span class="n">from_smi</span><span class="p">(</span><span class="n">s</span><span class="p">,</span><span class="n">y</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span><span class="p">,</span><span class="n">y</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">smis</span><span class="p">,</span><span class="n">ys</span><span class="p">)]</span>
<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="n">dp</span><span class="o">.</span><span class="n">mol</span> <span class="k">for</span> <span class="n">dp</span> <span class="ow">in</span> <span class="n">tox_dps</span><span class="p">]</span>
<span class="n">tr_idx</span><span class="p">,</span> <span class="n">va_idx</span><span class="p">,</span> <span class="n">te_idx</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">make_split_indices</span><span class="p">(</span><span class="n">mols</span><span class="p">,</span> <span class="s2">&quot;random&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mf">0.8</span><span class="p">,</span><span class="mf">0.1</span><span class="p">,</span><span class="mf">0.1</span><span class="p">))</span>
<span class="n">tr</span><span class="p">,</span> <span class="n">va</span><span class="p">,</span> <span class="n">te</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">split_data_by_indices</span><span class="p">(</span><span class="n">tox_dps</span><span class="p">,</span> <span class="n">tr_idx</span><span class="p">,</span> <span class="n">va_idx</span><span class="p">,</span> <span class="n">te_idx</span><span class="p">)</span>

<span class="n">feat</span> <span class="o">=</span> <span class="n">featurizers</span><span class="o">.</span><span class="n">SimpleMoleculeMolGraphFeaturizer</span><span class="p">()</span>
<span class="n">tox_tr</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MoleculeDataset</span><span class="p">(</span><span class="n">tr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>
<span class="n">tox_va</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MoleculeDataset</span><span class="p">(</span><span class="n">va</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>
<span class="n">tox_te</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MoleculeDataset</span><span class="p">(</span><span class="n">te</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="model-and-training">
<h3>4.2 Model and training<a class="headerlink" href="#model-and-training" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">mp</span>  <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">BondMessagePassing</span><span class="p">()</span>
<span class="n">agg</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">MeanAggregation</span><span class="p">()</span>
<span class="n">ffn</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">BinaryClassificationFFN</span><span class="p">(</span><span class="n">n_tasks</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">mpnn_tox</span> <span class="o">=</span> <span class="n">models</span><span class="o">.</span><span class="n">MPNN</span><span class="p">(</span><span class="n">mp</span><span class="p">,</span> <span class="n">agg</span><span class="p">,</span> <span class="n">ffn</span><span class="p">,</span> <span class="n">batch_norm</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>  

<span class="n">tr_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">tox_tr</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">va_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">tox_va</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">te_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">tox_te</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">trainer_tox</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">Trainer</span><span class="p">(</span><span class="n">logger</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">enable_checkpointing</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">accelerator</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span>
                         <span class="n">devices</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_epochs</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
<span class="n">trainer_tox</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">mpnn_tox</span><span class="p">,</span> <span class="n">tr_loader</span><span class="p">,</span> <span class="n">va_loader</span><span class="p">)</span>
<span class="n">trainer_tox</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">mpnn_tox</span><span class="p">,</span> <span class="n">te_loader</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition-output admonition">
<p class="admonition-title">Output</p>
<p>┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         test/roc          │    0.8480725288391113     │
└───────────────────────────┴───────────────────────────┘
[{‘test/roc’: 0.8480725288391113}]</p>
</div>
</section>
<section id="roc-curve">
<h3>4.3 ROC curve<a class="headerlink" href="#roc-curve" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>
<span class="c1"># Gather probabilities</span>
<span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">inference_mode</span><span class="p">():</span>
    <span class="n">pred_chunks</span> <span class="o">=</span> <span class="n">trainer_tox</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">mpnn_tox</span><span class="p">,</span> <span class="n">te_loader</span><span class="p">)</span>
<span class="n">proba</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">pred_chunks</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="n">y_true</span> <span class="o">=</span> <span class="n">tox_te</span><span class="o">.</span><span class="n">Y</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

<span class="n">auc</span> <span class="o">=</span> <span class="n">roc_auc_score</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">proba</span><span class="p">)</span>
<span class="n">acc</span> <span class="o">=</span> <span class="n">accuracy_score</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="p">(</span><span class="n">proba</span><span class="o">&gt;=</span><span class="mf">0.5</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Test AUC: </span><span class="si">{</span><span class="n">auc</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">  Accuracy: </span><span class="si">{</span><span class="n">acc</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">fpr</span><span class="p">,</span> <span class="n">tpr</span><span class="p">,</span> <span class="n">thr</span> <span class="o">=</span> <span class="n">roc_curve</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">proba</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">fpr</span><span class="p">,</span> <span class="n">tpr</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span><span class="s2">&quot;k--&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;FPR&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;TPR&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;ROC: Toxicity&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<div class="cell tag_hide-input docutils container">
<details class="hide above-input">
<summary aria-label="Toggle hidden content">
<span class="collapsed">Show code cell source</span>
<span class="expanded">Hide code cell source</span>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">IPython.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">Image</span><span class="p">,</span> <span class="n">display</span>
<span class="n">display</span><span class="p">(</span><span class="n">Image</span><span class="p">(</span><span class="n">url</span><span class="o">=</span><span class="s2">&quot;https://raw.githubusercontent.com/zzhenglab/ai4chem/main/book/_data/lec-10-roc.png&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<div class="output text_html"><img src="https://raw.githubusercontent.com/zzhenglab/ai4chem/main/book/_data/lec-10-roc.png"/></div></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="reactivity-and-selectivity">
<h2><a class="toc-backref" href="#id6" role="doc-backlink">5. Reactivity and selectivity</a><a class="headerlink" href="#reactivity-and-selectivity" title="Permalink to this heading">#</a></h2>
<p>Two targets in this dataset relate to reactions.</p>
<ul class="simple">
<li><p><strong>Reactivity</strong>: binary at the molecule level.</p></li>
<li><p><strong>Selectivity</strong>: oxidation site indices at the atom level.</p></li>
</ul>
<p>Before we train models, let’s <strong>look at the labels</strong>. Let’s pick three representative molecules from the C-H oxidation dataset and see their reactivity and selectivity label.</p>
<div class="cell tag_hide-input docutils container">
<details class="hide above-input">
<summary aria-label="Toggle hidden content">
<span class="collapsed">Show code cell source</span>
<span class="expanded">Hide code cell source</span>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Helper selection: one negative (no reaction), one positive with a single site, one positive with multiple sites</span>
<span class="k">def</span><span class="w"> </span><span class="nf">pick_representatives</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
    <span class="n">df_pos</span> <span class="o">=</span> <span class="n">df</span><span class="p">[(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;react_bin&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;site_list&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="nb">len</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">df_pos_multi</span> <span class="o">=</span> <span class="n">df_pos</span><span class="p">[</span><span class="n">df_pos</span><span class="p">[</span><span class="s2">&quot;site_list&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="nb">len</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">df_neg</span> <span class="o">=</span> <span class="n">df</span><span class="p">[(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;react_bin&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="n">reps</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">df_neg</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="n">reps</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="s2">&quot;Negative (react_bin=0)&quot;</span><span class="p">,</span> <span class="n">df_neg</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">df_pos</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="n">reps</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="s2">&quot;Positive (react_bin=1; 1 site)&quot;</span><span class="p">,</span> <span class="n">df_pos</span><span class="p">[</span><span class="n">df_pos</span><span class="p">[</span><span class="s2">&quot;site_list&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="nb">len</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                     <span class="k">if</span> <span class="p">(</span><span class="n">df_pos</span><span class="p">[</span><span class="s2">&quot;site_list&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="nb">len</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()</span> <span class="k">else</span> <span class="n">df_pos</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">df_pos_multi</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="n">reps</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="s2">&quot;Positive (react_bin=1; multi-site)&quot;</span><span class="p">,</span> <span class="n">df_pos_multi</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>

    <span class="c1"># If fewer than 3 examples exist, just return what we have</span>
    <span class="k">return</span> <span class="n">reps</span>

<span class="n">reps</span> <span class="o">=</span> <span class="n">pick_representatives</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
<span class="nb">len</span><span class="p">(</span><span class="n">reps</span><span class="p">),</span> <span class="p">[</span><span class="n">t</span> <span class="k">for</span> <span class="n">t</span><span class="p">,</span><span class="n">_</span> <span class="ow">in</span> <span class="n">reps</span><span class="p">]</span>
<span class="c1"># Show the chosen rows so readers see SMILES and labels</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="k">def</span><span class="w"> </span><span class="nf">row_view</span><span class="p">(</span><span class="n">r</span><span class="p">):</span>
    <span class="k">return</span> <span class="p">{</span>
        <span class="s2">&quot;SMILES&quot;</span><span class="p">:</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">],</span>
        <span class="s2">&quot;react_bin&quot;</span><span class="p">:</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;react_bin&quot;</span><span class="p">],</span>
        <span class="s2">&quot;site_list (1-based)&quot;</span><span class="p">:</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;site_list&quot;</span><span class="p">]</span>
    <span class="p">}</span>

<span class="n">rep_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">row_view</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">reps</span><span class="p">])</span>
<span class="n">rep_table</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>SMILES</th>
      <th>react_bin</th>
      <th>site_list (1-based)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Nc1c(Oc2ccccc2)cc(O)c2c1C(=O)c1ccccc1C2=O</td>
      <td>0</td>
      <td>[-1]</td>
    </tr>
    <tr>
      <th>1</th>
      <td>c1ccc2c(c1)Cc1ccccc1-2</td>
      <td>1</td>
      <td>[7]</td>
    </tr>
    <tr>
      <th>2</th>
      <td>c1ccc2c(c1)CCOC2</td>
      <td>1</td>
      <td>[8, 10]</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Let’s draw them:</p>
<div class="cell tag_hide-input docutils container">
<details class="hide above-input">
<summary aria-label="Toggle hidden content">
<span class="collapsed">Show code cell source</span>
<span class="expanded">Hide code cell source</span>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">rdkit.Chem.Draw</span><span class="w"> </span><span class="kn">import</span> <span class="n">rdMolDraw2D</span>

<span class="k">def</span><span class="w"> </span><span class="nf">make_annotated_copy</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">site_list_1based</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">tag_c123</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="n">m</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Mol</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>  <span class="c1"># copy</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">AssignAtomChiralTagsFromStructure</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">Kekulize</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">clearAromaticFlags</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">n</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">()</span>
    <span class="c1"># Highlight oxidation sites (convert to 0-based safely)</span>
    <span class="n">hi_atoms</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">site_list_1based</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">idx1</span> <span class="ow">in</span> <span class="n">site_list_1based</span><span class="p">:</span>
            <span class="n">j</span> <span class="o">=</span> <span class="n">idx1</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">n</span><span class="p">:</span>
                <span class="n">hi_atoms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">j</span><span class="p">)</span>

    <span class="c1"># Always annotate the atom index so readers see 1-based indexing used in labels</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">a</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">GetAtomWithIdx</span><span class="p">(</span><span class="n">j</span><span class="p">)</span>
        <span class="n">idx1</span> <span class="o">=</span> <span class="n">j</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">old</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">GetProp</span><span class="p">(</span><span class="s2">&quot;atomNote&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="n">a</span><span class="o">.</span><span class="n">HasProp</span><span class="p">(</span><span class="s2">&quot;atomNote&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
        <span class="c1"># If this atom is an oxidation site, add a star</span>
        <span class="n">star</span> <span class="o">=</span> <span class="s2">&quot;*&quot;</span> <span class="k">if</span> <span class="p">(</span><span class="n">j</span> <span class="ow">in</span> <span class="n">hi_atoms</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
        <span class="n">a</span><span class="o">.</span><span class="n">SetProp</span><span class="p">(</span><span class="s2">&quot;atomNote&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">old</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">idx1</span><span class="si">}{</span><span class="n">star</span><span class="si">}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">m</span><span class="p">,</span> <span class="n">hi_atoms</span>

<span class="k">def</span><span class="w"> </span><span class="nf">draw_examples</span><span class="p">(</span><span class="n">reps</span><span class="p">,</span> <span class="n">mol_size</span><span class="o">=</span><span class="p">(</span><span class="mi">320</span><span class="p">,</span> <span class="mi">280</span><span class="p">)):</span>
    <span class="n">ms</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">legends</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">highlights</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">title</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">reps</span><span class="p">:</span>
        <span class="n">smi</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">]</span>
        <span class="n">m</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">smi</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">m</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">m_annot</span><span class="p">,</span> <span class="n">hi</span> <span class="o">=</span> <span class="n">make_annotated_copy</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">site_list_1based</span><span class="o">=</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;site_list&quot;</span><span class="p">],</span> <span class="n">tag_c123</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">ms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">m_annot</span><span class="p">)</span>
        <span class="n">lbl</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Reactivity=</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;react_bin&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">, Sites=</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;site_list&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">legends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lbl</span><span class="p">)</span>
        <span class="n">highlights</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">hi</span><span class="p">)</span>

    <span class="n">imgs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">m</span><span class="p">,</span> <span class="n">hi</span><span class="p">,</span> <span class="n">lg</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">ms</span><span class="p">,</span> <span class="n">highlights</span><span class="p">,</span> <span class="n">legends</span><span class="p">):</span>
        <span class="n">img</span> <span class="o">=</span> <span class="n">Draw</span><span class="o">.</span><span class="n">MolToImage</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">mol_size</span><span class="p">,</span> <span class="n">highlightAtoms</span><span class="o">=</span><span class="n">hi</span><span class="p">)</span>
        <span class="n">img</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;legend&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">lg</span>
        <span class="n">imgs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">img</span><span class="p">)</span>

    <span class="c1"># Create a grid image manually by re-drawing with legends</span>
    <span class="k">return</span> <span class="n">Draw</span><span class="o">.</span><span class="n">MolsToGridImage</span><span class="p">(</span><span class="n">ms</span><span class="p">,</span> <span class="n">molsPerRow</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">ms</span><span class="p">),</span> <span class="n">subImgSize</span><span class="o">=</span><span class="n">mol_size</span><span class="p">,</span>
                                <span class="n">legends</span><span class="o">=</span><span class="n">legends</span><span class="p">,</span>
                                <span class="n">highlightAtomLists</span><span class="o">=</span><span class="n">highlights</span><span class="p">)</span>

<span class="n">grid_img</span> <span class="o">=</span> <span class="n">draw_examples</span><span class="p">(</span><span class="n">reps</span><span class="p">)</span>
<span class="n">grid_img</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<img alt="_images/cb6d2b1e39651169b4b8916f421bc50fac71c06ca93ac43edb69eb4baae2e041.png" src="_images/cb6d2b1e39651169b4b8916f421bc50fac71c06ca93ac43edb69eb4baae2e041.png" />
</div>
</div>
<section id="reactivity-classifier">
<h3>5.1 Reactivity classifier<a class="headerlink" href="#reactivity-classifier" title="Permalink to this heading">#</a></h3>
<p>This mirrors the toxicity workflow.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">df_rxn</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">,</span><span class="s2">&quot;react_bin&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
<span class="n">smis</span> <span class="o">=</span> <span class="n">df_rxn</span><span class="p">[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">ys</span>   <span class="o">=</span> <span class="n">df_rxn</span><span class="p">[</span><span class="s2">&quot;react_bin&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>

<span class="n">rxn_dps</span> <span class="o">=</span> <span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">MoleculeDatapoint</span><span class="o">.</span><span class="n">from_smi</span><span class="p">(</span><span class="n">s</span><span class="p">,</span><span class="n">y</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span><span class="p">,</span><span class="n">y</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">smis</span><span class="p">,</span> <span class="n">ys</span><span class="p">)]</span>
<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="n">dp</span><span class="o">.</span><span class="n">mol</span> <span class="k">for</span> <span class="n">dp</span> <span class="ow">in</span> <span class="n">rxn_dps</span><span class="p">]</span>
<span class="n">tr_idx</span><span class="p">,</span> <span class="n">va_idx</span><span class="p">,</span> <span class="n">te_idx</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">make_split_indices</span><span class="p">(</span><span class="n">mols</span><span class="p">,</span> <span class="s2">&quot;random&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mf">0.8</span><span class="p">,</span><span class="mf">0.1</span><span class="p">,</span><span class="mf">0.1</span><span class="p">))</span>
<span class="n">tr</span><span class="p">,</span> <span class="n">va</span><span class="p">,</span> <span class="n">te</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">split_data_by_indices</span><span class="p">(</span><span class="n">rxn_dps</span><span class="p">,</span> <span class="n">tr_idx</span><span class="p">,</span> <span class="n">va_idx</span><span class="p">,</span> <span class="n">te_idx</span><span class="p">)</span>

<span class="n">feat</span> <span class="o">=</span> <span class="n">featurizers</span><span class="o">.</span><span class="n">SimpleMoleculeMolGraphFeaturizer</span><span class="p">()</span>
<span class="n">rxn_tr</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MoleculeDataset</span><span class="p">(</span><span class="n">tr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>
<span class="n">rxn_va</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MoleculeDataset</span><span class="p">(</span><span class="n">va</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>
<span class="n">rxn_te</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MoleculeDataset</span><span class="p">(</span><span class="n">te</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>

<span class="n">mp</span>  <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">BondMessagePassing</span><span class="p">()</span>
<span class="n">agg</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">MeanAggregation</span><span class="p">()</span>
<span class="n">ffn</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">BinaryClassificationFFN</span><span class="p">(</span><span class="n">n_tasks</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">mpnn_rxn</span> <span class="o">=</span> <span class="n">models</span><span class="o">.</span><span class="n">MPNN</span><span class="p">(</span><span class="n">mp</span><span class="p">,</span> <span class="n">agg</span><span class="p">,</span> <span class="n">ffn</span><span class="p">,</span> <span class="n">batch_norm</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">tr_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">rxn_tr</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">va_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">rxn_va</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">te_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">rxn_te</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">trainer_rxn</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">Trainer</span><span class="p">(</span><span class="n">logger</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">enable_checkpointing</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">accelerator</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span>
                         <span class="n">devices</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_epochs</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
<span class="n">trainer_rxn</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">mpnn_rxn</span><span class="p">,</span> <span class="n">tr_loader</span><span class="p">,</span> <span class="n">va_loader</span><span class="p">)</span>
<span class="n">trainer_rxn</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">mpnn_rxn</span><span class="p">,</span> <span class="n">te_loader</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition-output admonition">
<p class="admonition-title">Output</p>
<p>[{‘test/roc’: 0.9569892883300781}]</p>
</div>
</section>
<section id="atom-level-selectivity-build-labels-per-atom">
<h3>5.2 Atom-level selectivity: build labels per atom<a class="headerlink" href="#atom-level-selectivity-build-labels-per-atom" title="Permalink to this heading">#</a></h3>
<p>We want atom targets <code class="docutils literal notranslate"><span class="pre">1</span></code> for indices that appear in <code class="docutils literal notranslate"><span class="pre">site_list</span></code>, and <code class="docutils literal notranslate"><span class="pre">0</span></code> otherwise.</p>
<p>We will create <strong>MolAtomBondDatapoint</strong> objects. For a gentle first pass, we only supply <code class="docutils literal notranslate"><span class="pre">atom_y</span></code> and leave other advanced inputs out.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>
<span class="k">def</span><span class="w"> </span><span class="nf">atoms_labels_from_smiles</span><span class="p">(</span><span class="n">smi</span><span class="p">,</span> <span class="n">positive_idxs</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">Chem</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="n">m</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">smi</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">m</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="n">n</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">()</span>
    <span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">positive_idxs</span><span class="p">:</span>
        <span class="c1"># dataset uses 1-based indexing in the text, RDKit uses 0-based</span>
        <span class="n">j</span> <span class="o">=</span> <span class="n">idx</span><span class="o">-</span><span class="mi">1</span>
        <span class="k">if</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">n</span><span class="p">:</span>
            <span class="n">y</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>
    <span class="k">return</span> <span class="n">y</span>

<span class="c1"># Build list of MolAtomBondDatapoint for selectivity</span>
<span class="n">sel_rows</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">,</span><span class="s2">&quot;site_list&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
<span class="n">sel_dps</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">smi</span><span class="p">,</span> <span class="n">sites</span> <span class="ow">in</span> <span class="n">sel_rows</span><span class="o">.</span><span class="n">itertuples</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="n">atom_y</span> <span class="o">=</span> <span class="n">atoms_labels_from_smiles</span><span class="p">(</span><span class="n">smi</span><span class="p">,</span> <span class="n">sites</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">atom_y</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">continue</span>
    <span class="c1"># We provide atom_y, molecule-level y is optional here</span>
    <span class="n">dp</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MolAtomBondDatapoint</span><span class="o">.</span><span class="n">from_smi</span><span class="p">(</span>
        <span class="n">smi</span><span class="p">,</span> <span class="n">atom_y</span><span class="o">=</span><span class="n">atom_y</span><span class="p">,</span> <span class="n">reorder_atoms</span><span class="o">=</span><span class="kc">False</span>
    <span class="p">)</span>
    <span class="n">sel_dps</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dp</span><span class="p">)</span>

<span class="nb">len</span><span class="p">(</span><span class="n">sel_dps</span><span class="p">),</span> <span class="nb">type</span><span class="p">(</span><span class="n">sel_dps</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="vm">__name__</span>
</pre></div>
</div>
<p>Split and dataset construction.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">dp</span><span class="o">.</span><span class="n">name</span><span class="p">)</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">dp</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="kc">None</span> <span class="k">for</span> <span class="n">dp</span> <span class="ow">in</span> <span class="n">sel_dps</span><span class="p">]</span>
<span class="c1"># For structure-based split we need RDKit Mol. Build directly from SMILES fallback:</span>
<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">]</span><span class="o">==</span><span class="n">dp</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s2">&quot;SMILES&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">if</span> <span class="n">Chem</span> <span class="k">else</span> <span class="kc">None</span> <span class="k">for</span> <span class="n">dp</span> <span class="ow">in</span> <span class="n">sel_dps</span><span class="p">]</span>

<span class="n">tr_idx</span><span class="p">,</span> <span class="n">va_idx</span><span class="p">,</span> <span class="n">te_idx</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">make_split_indices</span><span class="p">(</span><span class="n">mols</span><span class="p">,</span> <span class="s2">&quot;random&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">))</span>
<span class="n">tr</span><span class="p">,</span> <span class="n">va</span><span class="p">,</span> <span class="n">te</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">split_data_by_indices</span><span class="p">(</span><span class="n">sel_dps</span><span class="p">,</span> <span class="n">tr_idx</span><span class="p">,</span> <span class="n">va_idx</span><span class="p">,</span> <span class="n">te_idx</span><span class="p">)</span>

<span class="n">feat</span> <span class="o">=</span> <span class="n">featurizers</span><span class="o">.</span><span class="n">SimpleMoleculeMolGraphFeaturizer</span><span class="p">()</span>
<span class="n">tr_set</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MolAtomBondDataset</span><span class="p">(</span><span class="n">tr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>
<span class="n">va_set</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MolAtomBondDataset</span><span class="p">(</span><span class="n">va</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>
<span class="n">te_set</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MolAtomBondDataset</span><span class="p">(</span><span class="n">te</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>

<span class="n">tr_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">tr_set</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="n">va_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">va_set</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="n">te_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">te_set</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<p>Model for molecule and atom predictions. Here we focus on atom prediction.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">mp</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">MABBondMessagePassing</span><span class="p">(</span>
    <span class="n">d_v</span><span class="o">=</span><span class="n">feat</span><span class="o">.</span><span class="n">atom_fdim</span><span class="p">,</span> <span class="n">d_e</span><span class="o">=</span><span class="n">feat</span><span class="o">.</span><span class="n">bond_fdim</span><span class="p">,</span> <span class="n">d_h</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">depth</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">dropout</span><span class="o">=</span><span class="mf">0.1</span>
<span class="p">)</span>
<span class="n">agg</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">MeanAggregation</span><span class="p">()</span>

<span class="n">atom_predictor</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">BinaryClassificationFFN</span><span class="p">(</span><span class="n">n_tasks</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># atom-level 0/1</span>

<span class="n">model_sel</span> <span class="o">=</span> <span class="n">models</span><span class="o">.</span><span class="n">MolAtomBondMPNN</span><span class="p">(</span>
    <span class="n">message_passing</span><span class="o">=</span><span class="n">mp</span><span class="p">,</span>
    <span class="n">agg</span><span class="o">=</span><span class="n">agg</span><span class="p">,</span>
    <span class="n">mol_predictor</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">atom_predictor</span><span class="o">=</span><span class="n">atom_predictor</span><span class="p">,</span>
    <span class="n">bond_predictor</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">batch_norm</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">metrics</span><span class="o">=</span><span class="p">[</span><span class="n">nn</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">BinaryAUROC</span><span class="p">()],</span>
<span class="p">)</span>

<span class="n">trainer_sel</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">Trainer</span><span class="p">(</span><span class="n">logger</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">enable_checkpointing</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">accelerator</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span>
                         <span class="n">devices</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_epochs</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="n">trainer_sel</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">model_sel</span><span class="p">,</span> <span class="n">tr_loader</span><span class="p">,</span> <span class="n">va_loader</span><span class="p">)</span>
<span class="n">trainer_sel</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">model_sel</span><span class="p">,</span> <span class="n">te_loader</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition-output admonition">
<p class="admonition-title">Output</p>
<p>[{‘atom_test/roc’: 0.8910984396934509}]</p>
</div>
<p>Inspect a molecule with predicted atom probabilities by Visualizing probabilities as atom annotations for the first molecule in the batch.</p>
<div class="highlight-PYTHON notranslate"><div class="highlight"><pre><span></span><span class="p">:</span><span class="n">tags</span><span class="p">:</span> <span class="p">[</span><span class="n">hide</span><span class="o">-</span><span class="nb">input</span><span class="p">]</span>
<span class="c1"># 1) Pick which test item to visualize</span>
<span class="n">idx</span> <span class="o">=</span> <span class="mi">11</span>  <span class="c1"># change as you like</span>

<span class="c1"># 2) Grab the original datapoint (not the already-featurized datum)</span>
<span class="n">dp</span> <span class="o">=</span> <span class="n">te_set</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>   <span class="c1"># IMPORTANT: .data gives you the raw MolAtomBondDatapoint</span>

<span class="c1"># 3) Make a one-item MolAtomBondDataset with the SAME featurizer you used before</span>
<span class="n">single_ds</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MolAtomBondDataset</span><span class="p">([</span><span class="n">dp</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>

<span class="c1"># 4) Build a loader and get its batch</span>
<span class="n">single_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">single_ds</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">batch</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">single_loader</span><span class="p">))</span>  <span class="c1"># this now matches collate expectations</span>

<span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">inference_mode</span><span class="p">():</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">model_sel</span><span class="p">(</span><span class="n">batch</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="kc">None</span><span class="p">)</span>   <span class="c1"># batch[0] is the MolGraphBatch</span>

<span class="c1"># 6) Unpack outputs and get per-atom probabilities</span>
<span class="n">_</span><span class="p">,</span> <span class="n">atom_logits</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">out</span>
<span class="n">atom_probs</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">sigmoid</span><span class="p">(</span><span class="n">atom_logits</span><span class="p">)</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Atom count:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">atom_probs</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;First 10 probabilities:&quot;</span><span class="p">,</span> <span class="n">atom_probs</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>

<span class="c1"># 7) Draw the SAME molecule with aligned probabilities</span>
<span class="n">smi</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">name</span>  <span class="c1"># MolAtomBondDatapoint stores SMILES in .name</span>
<span class="n">m</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">smi</span><span class="p">)</span>
<span class="n">m2</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Mol</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">m2</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">()):</span>
    <span class="n">a</span><span class="o">.</span><span class="n">SetProp</span><span class="p">(</span><span class="s2">&quot;atomNote&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">atom_probs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">img</span> <span class="o">=</span> <span class="n">Draw</span><span class="o">.</span><span class="n">MolToImage</span><span class="p">(</span><span class="n">m2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">400</span><span class="p">,</span> <span class="mi">400</span><span class="p">))</span>
<span class="n">display</span><span class="p">(</span><span class="n">img</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition-exercise-7-1 admonition">
<p class="admonition-title">⏰ Exercise 7.1</p>
<p>Switch the atom predictor to a small regression head and train with labels 0.0 or 1.0. Then threshold the outputs at 0.5 to recover classes. Compare AUROC.</p>
</div>
</section>
</section>
<hr class="docutils" />
<section id="rationale">
<h2><a class="toc-backref" href="#id7" role="doc-backlink">6. Rationale</a><a class="headerlink" href="#rationale" title="Permalink to this heading">#</a></h2>
<p>In this section we ask <strong>why</strong> the model makes its predictions. Instead of just reporting a number for solubility or reactivity, we try to find the <em>substructures</em> of a molecule that drive the prediction.</p>
<p>The method we use is a simplified form of <strong>Monte Carlo Tree Search (MCTS) rationale extraction</strong> from the Chemprop:</p>
<ol class="arabic simple">
<li><p><strong>Start with the whole molecule</strong> and its predicted property.</p></li>
<li><p><strong>Systematically remove small fragments</strong> (like a bond or a ring) and check how the model’s prediction changes for the remaining subgraph.</p></li>
<li><p>Build a search tree of candidate subgraphs and keep the ones that still predict strongly for the property of interest.</p></li>
<li><p>From this set, report the <strong>smallest and most predictive subgraph</strong> as the <em>rationale</em>.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>
<span class="c1"># Predict a list of SMILES with a single-task model. Returns array shape (n, 1).</span>
<span class="k">def</span><span class="w"> </span><span class="nf">make_prediction</span><span class="p">(</span><span class="n">models_in</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">MPNN</span><span class="p">],</span> <span class="n">trainer_in</span><span class="p">:</span> <span class="n">pl</span><span class="o">.</span><span class="n">Trainer</span><span class="p">,</span> <span class="n">smiles</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
    <span class="n">test_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">MoleculeDatapoint</span><span class="o">.</span><span class="n">from_smi</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">smiles</span><span class="p">]</span>
    <span class="n">test_dset</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MoleculeDataset</span><span class="p">(</span><span class="n">test_data</span><span class="p">)</span>
    <span class="n">test_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">test_dset</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">inference_mode</span><span class="p">():</span>
        <span class="n">agg</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">models_in</span><span class="p">:</span>
            <span class="n">pred_batches</span> <span class="o">=</span> <span class="n">trainer_in</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">test_loader</span><span class="p">)</span>  <span class="c1"># list of tensors/arrays</span>
            <span class="n">preds</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cat</span><span class="p">([</span><span class="n">torch</span><span class="o">.</span><span class="n">as_tensor</span><span class="p">(</span><span class="n">pb</span><span class="p">)</span> <span class="k">for</span> <span class="n">pb</span> <span class="ow">in</span> <span class="n">pred_batches</span><span class="p">],</span> <span class="n">dim</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
            <span class="n">agg</span> <span class="o">=</span> <span class="n">preds</span> <span class="k">if</span> <span class="n">agg</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="p">(</span><span class="n">agg</span> <span class="o">+</span> <span class="n">preds</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">agg</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">models_in</span><span class="p">)</span>

<span class="c1"># ---- MCTS components (as in tutorial, with safer kekulize handling) ----</span>
<span class="nd">@dataclass</span>
<span class="k">class</span><span class="w"> </span><span class="nc">MCTSNode</span><span class="p">:</span>
    <span class="n">smiles</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">atoms</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span>
    <span class="n">W</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">N</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">P</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">children</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="n">field</span><span class="p">(</span><span class="n">default_factory</span><span class="o">=</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">__post_init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span> <span class="bp">self</span><span class="o">.</span><span class="n">atoms</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">atoms</span><span class="p">)</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">Q</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span> <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">W</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">N</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">N</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">U</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">c_puct</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">10.0</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span> <span class="k">return</span> <span class="n">c_puct</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">P</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">N</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">find_clusters</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Mol</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span><span class="o">...</span><span class="p">]],</span> <span class="n">List</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]]]:</span>
    <span class="n">n</span> <span class="o">=</span> <span class="n">mol</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="k">return</span> <span class="p">[(</span><span class="mi">0</span><span class="p">,)],</span> <span class="p">[[</span><span class="mi">0</span><span class="p">]]</span>
    <span class="n">clusters</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">mol</span><span class="o">.</span><span class="n">GetBonds</span><span class="p">():</span>
        <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="o">=</span> <span class="n">b</span><span class="o">.</span><span class="n">GetBeginAtomIdx</span><span class="p">(),</span> <span class="n">b</span><span class="o">.</span><span class="n">GetEndAtomIdx</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">b</span><span class="o">.</span><span class="n">IsInRing</span><span class="p">():</span>
            <span class="n">clusters</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">))</span>
    <span class="n">ssr</span> <span class="o">=</span> <span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">Chem</span><span class="o">.</span><span class="n">GetSymmSSSR</span><span class="p">(</span><span class="n">mol</span><span class="p">)]</span>
    <span class="n">clusters</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">ssr</span><span class="p">)</span>
    <span class="n">atom_cls</span> <span class="o">=</span> <span class="p">[[]</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">)]</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">cl</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">clusters</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">cl</span><span class="p">:</span> <span class="n">atom_cls</span><span class="p">[</span><span class="n">a</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">clusters</span><span class="p">,</span> <span class="n">atom_cls</span>

<span class="k">def</span><span class="w"> </span><span class="nf">extract_subgraph_from_mol</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Mol</span><span class="p">,</span> <span class="n">selected_atoms</span><span class="p">:</span> <span class="nb">set</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Chem</span><span class="o">.</span><span class="n">Mol</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]]:</span>
    <span class="n">sel</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">selected_atoms</span><span class="p">)</span>
    <span class="n">roots</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">sel</span><span class="p">:</span>
        <span class="n">atom</span> <span class="o">=</span> <span class="n">mol</span><span class="o">.</span><span class="n">GetAtomWithIdx</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">nei</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sel</span> <span class="k">for</span> <span class="n">nei</span> <span class="ow">in</span> <span class="n">atom</span><span class="o">.</span><span class="n">GetNeighbors</span><span class="p">()):</span>
            <span class="n">roots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
    <span class="n">rw</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">RWMol</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">roots</span><span class="p">:</span>
        <span class="n">a</span> <span class="o">=</span> <span class="n">rw</span><span class="o">.</span><span class="n">GetAtomWithIdx</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
        <span class="n">a</span><span class="o">.</span><span class="n">SetAtomMapNum</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">aroma</span> <span class="o">=</span> <span class="p">[</span><span class="n">b</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">a</span><span class="o">.</span><span class="n">GetBonds</span><span class="p">()</span> <span class="k">if</span> <span class="n">b</span><span class="o">.</span><span class="n">GetBondType</span><span class="p">()</span> <span class="o">==</span> <span class="n">Chem</span><span class="o">.</span><span class="n">rdchem</span><span class="o">.</span><span class="n">BondType</span><span class="o">.</span><span class="n">AROMATIC</span><span class="p">]</span>
        <span class="n">aroma</span> <span class="o">=</span> <span class="p">[</span><span class="n">b</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">aroma</span> <span class="k">if</span> <span class="n">b</span><span class="o">.</span><span class="n">GetBeginAtom</span><span class="p">()</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span> <span class="ow">in</span> <span class="n">sel</span> <span class="ow">and</span> <span class="n">b</span><span class="o">.</span><span class="n">GetEndAtom</span><span class="p">()</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span> <span class="ow">in</span> <span class="n">sel</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">aroma</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="n">a</span><span class="o">.</span><span class="n">SetIsAromatic</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">a</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">rw</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">()</span> <span class="k">if</span> <span class="n">a</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sel</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="n">rw</span><span class="o">.</span><span class="n">RemoveAtom</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">rw</span><span class="o">.</span><span class="n">GetMol</span><span class="p">(),</span> <span class="n">roots</span>

<span class="k">def</span><span class="w"> </span><span class="nf">extract_subgraph</span><span class="p">(</span><span class="n">smiles</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">selected_atoms</span><span class="p">:</span> <span class="nb">set</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="kc">None</span><span class="p">],</span> <span class="n">Union</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span><span class="kc">None</span><span class="p">]]:</span>
    <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">smiles</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span> <span class="k">return</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span>
    <span class="c1"># try kekulized</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">mk</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Mol</span><span class="p">(</span><span class="n">mol</span><span class="p">);</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Kekulize</span><span class="p">(</span><span class="n">mk</span><span class="p">)</span>
        <span class="n">sub</span><span class="p">,</span> <span class="n">roots</span> <span class="o">=</span> <span class="n">extract_subgraph_from_mol</span><span class="p">(</span><span class="n">mk</span><span class="p">,</span> <span class="n">selected_atoms</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">smi</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolToSmiles</span><span class="p">(</span><span class="n">sub</span><span class="p">,</span> <span class="n">kekuleSmiles</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">sub</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">smi</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span> <span class="n">sub</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">sub</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">mol</span><span class="o">.</span><span class="n">HasSubstructMatch</span><span class="p">(</span><span class="n">sub</span><span class="p">):</span> <span class="k">return</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolToSmiles</span><span class="p">(</span><span class="n">sub</span><span class="p">),</span> <span class="n">roots</span>
    <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="c1"># fallback without kekulize</span>
    <span class="n">sub</span><span class="p">,</span> <span class="n">roots</span> <span class="o">=</span> <span class="n">extract_subgraph_from_mol</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">selected_atoms</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">smi</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolToSmiles</span><span class="p">(</span><span class="n">sub</span><span class="p">)</span>
        <span class="n">sub</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">smi</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span> <span class="n">sub</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">MolToSmiles</span><span class="p">(</span><span class="n">sub</span><span class="p">),</span> <span class="n">roots</span><span class="p">)</span> <span class="k">if</span> <span class="n">sub</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">mcts_rollout</span><span class="p">(</span><span class="n">node</span><span class="p">:</span> <span class="n">MCTSNode</span><span class="p">,</span> <span class="n">state_map</span><span class="p">,</span> <span class="n">orig_smiles</span><span class="p">,</span> <span class="n">clusters</span><span class="p">,</span> <span class="n">atom_cls</span><span class="p">,</span> <span class="n">nei_cls</span><span class="p">,</span>
                 <span class="n">scoring_fn</span><span class="p">:</span> <span class="n">Callable</span><span class="p">[[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> <span class="n">min_atoms</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">c_puct</span><span class="o">=</span><span class="mf">10.0</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
    <span class="n">cur</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">atoms</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cur</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">min_atoms</span><span class="p">:</span> <span class="k">return</span> <span class="n">node</span><span class="o">.</span><span class="n">P</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">cur_cls</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">clusters</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span> <span class="o">&lt;=</span> <span class="n">cur</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">cur_cls</span><span class="p">:</span>
            <span class="n">leaf_atoms</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">clusters</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">atom_cls</span><span class="p">[</span><span class="n">a</span><span class="p">]</span> <span class="o">&amp;</span> <span class="n">cur_cls</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">nei_cls</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&amp;</span> <span class="n">cur_cls</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">clusters</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">==</span> <span class="mi">2</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">leaf_atoms</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">):</span>
                <span class="n">new_atoms</span> <span class="o">=</span> <span class="n">cur</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">leaf_atoms</span><span class="p">)</span>
                <span class="n">new_smi</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">extract_subgraph</span><span class="p">(</span><span class="n">orig_smiles</span><span class="p">,</span> <span class="n">new_atoms</span><span class="p">)</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">new_smi</span><span class="p">:</span> <span class="k">continue</span>
                <span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">state_map</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">new_smi</span><span class="p">,</span> <span class="n">MCTSNode</span><span class="p">(</span><span class="n">new_smi</span><span class="p">,</span> <span class="n">new_atoms</span><span class="p">)))</span>
        <span class="n">state_map</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">smiles</span><span class="p">]</span> <span class="o">=</span> <span class="n">node</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="k">return</span> <span class="n">node</span><span class="o">.</span><span class="n">P</span>
        <span class="n">scores</span> <span class="o">=</span> <span class="n">scoring_fn</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">smiles</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">child</span><span class="p">,</span> <span class="n">sc</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="p">,</span> <span class="n">scores</span><span class="p">):</span> <span class="n">child</span><span class="o">.</span><span class="n">P</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">sc</span><span class="p">)</span>
    <span class="n">totalN</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">c</span><span class="o">.</span><span class="n">N</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="p">)</span>
    <span class="n">nxt</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">Q</span><span class="p">()</span> <span class="o">+</span> <span class="n">x</span><span class="o">.</span><span class="n">U</span><span class="p">(</span><span class="n">totalN</span><span class="p">,</span> <span class="n">c_puct</span><span class="o">=</span><span class="n">c_puct</span><span class="p">))</span>
    <span class="n">v</span> <span class="o">=</span> <span class="n">mcts_rollout</span><span class="p">(</span><span class="n">nxt</span><span class="p">,</span> <span class="n">state_map</span><span class="p">,</span> <span class="n">orig_smiles</span><span class="p">,</span> <span class="n">clusters</span><span class="p">,</span> <span class="n">atom_cls</span><span class="p">,</span> <span class="n">nei_cls</span><span class="p">,</span> <span class="n">scoring_fn</span><span class="p">,</span> <span class="n">min_atoms</span><span class="o">=</span><span class="n">min_atoms</span><span class="p">,</span> <span class="n">c_puct</span><span class="o">=</span><span class="n">c_puct</span><span class="p">)</span>
    <span class="n">nxt</span><span class="o">.</span><span class="n">W</span> <span class="o">+=</span> <span class="n">v</span><span class="p">;</span> <span class="n">nxt</span><span class="o">.</span><span class="n">N</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">v</span>

<span class="k">def</span><span class="w"> </span><span class="nf">run_mcts_for_smiles</span><span class="p">(</span><span class="n">smi</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">scoring_fn</span><span class="p">,</span> <span class="n">rollout</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">c_puct</span><span class="o">=</span><span class="mf">10.0</span><span class="p">,</span>
                        <span class="n">min_atoms</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">max_atoms</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">MCTSNode</span><span class="p">]:</span>
    <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">smi</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span> <span class="k">return</span> <span class="p">[]</span>
    <span class="n">clusters_raw</span><span class="p">,</span> <span class="n">atom_cls_raw</span> <span class="o">=</span> <span class="n">find_clusters</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
    <span class="n">clusters</span> <span class="o">=</span> <span class="p">[</span><span class="nb">set</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span> <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="n">clusters_raw</span><span class="p">]</span>
    <span class="n">atom_cls</span> <span class="o">=</span> <span class="p">[</span><span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">atom_cls_raw</span><span class="p">]</span>
    <span class="n">nei_cls</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">cl</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">clusters</span><span class="p">):</span>
        <span class="n">neigh</span> <span class="o">=</span> <span class="p">[</span><span class="n">nei</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">cl</span> <span class="k">for</span> <span class="n">nei</span> <span class="ow">in</span> <span class="n">atom_cls_raw</span><span class="p">[</span><span class="n">a</span><span class="p">]]</span>
        <span class="n">nei_cls</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">neigh</span><span class="p">)</span> <span class="o">-</span> <span class="p">{</span><span class="n">i</span><span class="p">})</span>
    <span class="n">root</span> <span class="o">=</span> <span class="n">MCTSNode</span><span class="p">(</span><span class="n">smi</span><span class="p">,</span> <span class="nb">set</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">mol</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">())))</span>
    <span class="n">state_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">smi</span><span class="p">:</span> <span class="n">root</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">rollout</span><span class="p">):</span>
        <span class="n">mcts_rollout</span><span class="p">(</span><span class="n">root</span><span class="p">,</span> <span class="n">state_map</span><span class="p">,</span> <span class="n">smi</span><span class="p">,</span> <span class="n">clusters</span><span class="p">,</span> <span class="n">atom_cls</span><span class="p">,</span> <span class="n">nei_cls</span><span class="p">,</span> <span class="n">scoring_fn</span><span class="p">,</span> <span class="n">min_atoms</span><span class="o">=</span><span class="n">min_atoms</span><span class="p">,</span> <span class="n">c_puct</span><span class="o">=</span><span class="n">c_puct</span><span class="p">)</span>
    <span class="n">rats</span> <span class="o">=</span> <span class="p">[</span><span class="n">node</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">state_map</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">smiles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">atoms</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">max_atoms</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">rats</span>

<span class="c1"># Simple fragment fallback if MCTS cannot find anything</span>
<span class="k">def</span><span class="w"> </span><span class="nf">find_fragments</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Mol</span><span class="p">):</span>
    <span class="n">frags</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">mol</span><span class="o">.</span><span class="n">GetBonds</span><span class="p">():</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">b</span><span class="o">.</span><span class="n">IsInRing</span><span class="p">():</span>
            <span class="n">frags</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="n">b</span><span class="o">.</span><span class="n">GetBeginAtomIdx</span><span class="p">(),</span> <span class="n">b</span><span class="o">.</span><span class="n">GetEndAtomIdx</span><span class="p">()})</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ring</span> <span class="ow">in</span> <span class="n">Chem</span><span class="o">.</span><span class="n">GetSymmSSSR</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
            <span class="n">frags</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ring</span><span class="p">))</span>
    <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">return</span> <span class="n">frags</span>

<span class="k">def</span><span class="w"> </span><span class="nf">fragment_top1</span><span class="p">(</span><span class="n">smi</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">scoring_fn</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]:</span>
    <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">smi</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span> <span class="k">return</span> <span class="p">[]</span>
    <span class="n">cands</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">aset</span> <span class="ow">in</span> <span class="n">find_fragments</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
        <span class="n">sub_smi</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">extract_subgraph</span><span class="p">(</span><span class="n">smi</span><span class="p">,</span> <span class="n">aset</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">sub_smi</span><span class="p">:</span> <span class="n">cands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sub_smi</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">cands</span><span class="p">:</span> <span class="k">return</span> <span class="p">[]</span>
    <span class="n">cands</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">dict</span><span class="o">.</span><span class="n">fromkeys</span><span class="p">(</span><span class="n">cands</span><span class="p">))</span>
    <span class="n">scores</span> <span class="o">=</span> <span class="n">scoring_fn</span><span class="p">(</span><span class="n">cands</span><span class="p">)</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">scores</span><span class="p">))</span>
    <span class="k">return</span> <span class="p">[(</span><span class="n">cands</span><span class="p">[</span><span class="n">idx</span><span class="p">],</span> <span class="nb">float</span><span class="p">(</span><span class="n">scores</span><span class="p">[</span><span class="n">idx</span><span class="p">]))]</span>

<span class="k">def</span><span class="w"> </span><span class="nf">visualize_rationale_on_parent</span><span class="p">(</span><span class="n">parent_smi</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">rationale_smi</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">520</span><span class="p">,</span> <span class="mi">420</span><span class="p">)):</span>
    <span class="n">pm</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">parent_smi</span><span class="p">)</span>
    <span class="n">rm</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">rationale_smi</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">pm</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">rm</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Cannot draw: invalid SMILES.&quot;</span><span class="p">)</span>
        <span class="k">return</span>
    <span class="n">match</span> <span class="o">=</span> <span class="n">pm</span><span class="o">.</span><span class="n">GetSubstructMatch</span><span class="p">(</span><span class="n">rm</span><span class="p">)</span>
    <span class="n">hi_atoms</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">match</span><span class="p">)</span> <span class="k">if</span> <span class="n">match</span> <span class="k">else</span> <span class="p">[]</span>
    <span class="n">drawer</span> <span class="o">=</span> <span class="n">rdMolDraw2D</span><span class="o">.</span><span class="n">MolDraw2DCairo</span><span class="p">(</span><span class="n">size</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">size</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">opts</span> <span class="o">=</span> <span class="n">drawer</span><span class="o">.</span><span class="n">drawOptions</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">attr</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;atomLabelFontSize&quot;</span><span class="p">,</span> <span class="s2">&quot;fontSize&quot;</span><span class="p">,</span> <span class="s2">&quot;scalingFactor&quot;</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">opts</span><span class="p">,</span> <span class="n">attr</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="nb">setattr</span><span class="p">(</span><span class="n">opts</span><span class="p">,</span> <span class="n">attr</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">)</span>
                <span class="k">break</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">pass</span>
    <span class="n">rdMolDraw2D</span><span class="o">.</span><span class="n">PrepareAndDrawMolecule</span><span class="p">(</span><span class="n">drawer</span><span class="p">,</span> <span class="n">pm</span><span class="p">,</span>
                                       <span class="n">highlightAtoms</span><span class="o">=</span><span class="n">hi_atoms</span> <span class="k">if</span> <span class="n">hi_atoms</span> <span class="k">else</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">drawer</span><span class="o">.</span><span class="n">FinishDrawing</span><span class="p">()</span>
    <span class="n">png</span> <span class="o">=</span> <span class="n">drawer</span><span class="o">.</span><span class="n">GetDrawingText</span><span class="p">()</span>
    <span class="n">display</span><span class="p">(</span><span class="n">Image</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">png</span><span class="p">))</span>

</pre></div>
</div>
<p>Because scoring is done by predicting each subgraph as if it were a standalone molecule, the absolute values (like negative <code class="docutils literal notranslate"><span class="pre">logS</span></code>) may not be directly comparable to the parent’s score. What matters is the <strong>difference (Δ)</strong>: how much the rationale’s score differs from the parent. A positive Δ means the subgraph supports the property; a negative Δ means removing that part lowers the property.</p>
<p>We generate a <strong>summary table</strong> with one rationale (<code class="docutils literal notranslate"><span class="pre">rationale_0</span></code>) for each molecule:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">smiles</span></code>: the original molecule</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">logS</span></code> or <code class="docutils literal notranslate"><span class="pre">react_prob</span></code>: the model’s prediction on the full molecule</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rationale_0</span></code>: the SMILES string of the key substructure</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rationale_0_score</span></code>: the model’s prediction on that substructure</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rationale_0_delta</span></code>: the difference between rationale and parent</p></li>
</ul>
<p>Finally, we <strong>visualize the rationale</strong> by highlighting the matching atoms in the parent molecule. This lets us</p>
<section id="solubility-prediction-rationale">
<h3>6.1 Solubility prediction rationale<a class="headerlink" href="#solubility-prediction-rationale" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># ===== Solubility summary table with top-1 rationale =====</span>
<span class="c1"># Uses your trained mpnn_sol and trainer, and df_sol[&quot;SMILES&quot;]</span>

<span class="n">models_sol</span> <span class="o">=</span> <span class="p">[</span><span class="n">mpnn_sol</span><span class="p">]</span>  <span class="c1"># must be single-task regression</span>
<span class="n">prop_name</span> <span class="o">=</span> <span class="s2">&quot;logS&quot;</span>

<span class="k">def</span><span class="w"> </span><span class="nf">scoring_fn_sol</span><span class="p">(</span><span class="n">smiles_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
    <span class="k">return</span> <span class="n">make_prediction</span><span class="p">(</span><span class="n">models_sol</span><span class="p">,</span> <span class="n">trainer</span><span class="p">,</span> <span class="n">smiles_list</span><span class="p">)[:,</span> <span class="mi">0</span><span class="p">]</span>

<span class="c1"># Build table</span>
<span class="n">N</span> <span class="o">=</span> <span class="mi">15</span>  <span class="c1"># how many molecules to summarize</span>
<span class="n">sample_smis</span> <span class="o">=</span> <span class="n">df_sol</span><span class="p">[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()[:</span><span class="n">N</span><span class="p">]</span>

<span class="n">rows</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;smiles&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="n">prop_name</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;rationale_0&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;rationale_0_score&quot;</span><span class="p">:</span> <span class="p">[]}</span>

<span class="n">t0</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
<span class="k">for</span> <span class="n">smi</span> <span class="ow">in</span> <span class="n">sample_smis</span><span class="p">:</span>
    <span class="n">base</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">scoring_fn_sol</span><span class="p">([</span><span class="n">smi</span><span class="p">])[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">rows</span><span class="p">[</span><span class="s2">&quot;smiles&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">smi</span><span class="p">)</span>
    <span class="n">rows</span><span class="p">[</span><span class="n">prop_name</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">base</span><span class="p">)</span>

    <span class="c1"># size-aware MCTS settings</span>
    <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">smi</span><span class="p">)</span>
    <span class="n">n</span> <span class="o">=</span> <span class="n">mol</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">()</span> <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="mi">20</span>
    <span class="n">min_atoms</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="mf">0.30</span> <span class="o">*</span> <span class="n">n</span><span class="p">))</span>
    <span class="n">max_atoms</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">min_atoms</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="mf">0.60</span> <span class="o">*</span> <span class="n">n</span><span class="p">))</span>

    <span class="n">rats</span> <span class="o">=</span> <span class="n">run_mcts_for_smiles</span><span class="p">(</span><span class="n">smi</span><span class="p">,</span> <span class="n">scoring_fn_sol</span><span class="p">,</span> <span class="n">rollout</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">c_puct</span><span class="o">=</span><span class="mf">10.0</span><span class="p">,</span>
                               <span class="n">min_atoms</span><span class="o">=</span><span class="n">min_atoms</span><span class="p">,</span> <span class="n">max_atoms</span><span class="o">=</span><span class="n">max_atoms</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">rats</span><span class="p">:</span>
        <span class="c1"># choose smallest subgraphs, then highest score</span>
        <span class="n">ms</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">atoms</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">rats</span><span class="p">)</span>
        <span class="n">kept</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">rats</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">atoms</span><span class="p">)</span> <span class="o">==</span> <span class="n">ms</span> <span class="ow">and</span> <span class="n">x</span><span class="o">.</span><span class="n">smiles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">]</span>
        <span class="n">kept</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">P</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">r0</span> <span class="o">=</span> <span class="n">kept</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">smiles</span>
        <span class="n">r0s</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">kept</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">P</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># fallback to fragment top1</span>
        <span class="n">top1</span> <span class="o">=</span> <span class="n">fragment_top1</span><span class="p">(</span><span class="n">smi</span><span class="p">,</span> <span class="n">scoring_fn_sol</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">top1</span><span class="p">:</span>
            <span class="n">r0</span><span class="p">,</span> <span class="n">r0s</span> <span class="o">=</span> <span class="n">top1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">r0</span><span class="p">,</span> <span class="n">r0s</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span>

    <span class="n">rows</span><span class="p">[</span><span class="s2">&quot;rationale_0&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r0</span><span class="p">)</span>
    <span class="n">rows</span><span class="p">[</span><span class="s2">&quot;rationale_0_score&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r0s</span><span class="p">)</span>

<span class="n">elapsed</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">t0</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Solubility: built table for </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">sample_smis</span><span class="p">)</span><span class="si">}</span><span class="s2"> molecules in </span><span class="si">{</span><span class="n">elapsed</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">s&quot;</span><span class="p">)</span>



<span class="n">solubility_rationales</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">rows</span><span class="p">)</span>
<span class="n">solubility_rationales</span>

<span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notna</span><span class="p">(</span><span class="n">solubility_rationales</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;rationale_0&quot;</span><span class="p">]):</span>
    <span class="n">parent</span> <span class="o">=</span> <span class="n">solubility_rationales</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;smiles&quot;</span><span class="p">]</span>
    <span class="n">rat</span>    <span class="o">=</span> <span class="n">solubility_rationales</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;rationale_0&quot;</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Parent:&quot;</span><span class="p">,</span> <span class="n">parent</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Rationale 0:&quot;</span><span class="p">,</span> <span class="n">rat</span><span class="p">,</span> <span class="s2">&quot; score=&quot;</span><span class="p">,</span> <span class="n">solubility_rationales</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;rationale_0_score&quot;</span><span class="p">])</span>
    <span class="n">visualize_rationale_on_parent</span><span class="p">(</span><span class="n">parent</span><span class="p">,</span> <span class="n">rat</span><span class="p">)</span>

</pre></div>
</div>
</section>
<section id="reactivity-prediction-rationale">
<h3>6.2 Reactivity prediction rationale<a class="headerlink" href="#reactivity-prediction-rationale" title="Permalink to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># ===== Reactivity summary table with top-1 rationale =====</span>
<span class="c1"># Uses your trained mpnn_rxn and trainer_rxn, and df_rxn[&quot;SMILES&quot;]</span>

<span class="n">models_rxn</span> <span class="o">=</span> <span class="p">[</span><span class="n">mpnn_rxn</span><span class="p">]</span>     <span class="c1"># single-task binary classification head</span>
<span class="n">trainer_cls</span> <span class="o">=</span> <span class="n">trainer_rxn</span>   <span class="c1"># your classification trainer</span>

<span class="k">def</span><span class="w"> </span><span class="nf">scoring_fn_rxn</span><span class="p">(</span><span class="n">smiles_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
    <span class="c1"># predict logits, convert to probabilities with sigmoid</span>
    <span class="n">preds</span> <span class="o">=</span> <span class="n">make_prediction</span><span class="p">(</span><span class="n">models_rxn</span><span class="p">,</span> <span class="n">trainer_cls</span><span class="p">,</span> <span class="n">smiles_list</span><span class="p">)[:,</span> <span class="mi">0</span><span class="p">]</span>
    <span class="c1"># Some builds return probs already; to be safe, apply sigmoid if outside [0,1]</span>
    <span class="k">if</span> <span class="p">((</span><span class="n">preds</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">preds</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
        <span class="n">preds</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">preds</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">preds</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,)</span>

<span class="c1"># Build table</span>
<span class="n">N</span> <span class="o">=</span> <span class="mi">15</span>
<span class="n">sample_smis_rxn</span> <span class="o">=</span> <span class="n">df_rxn</span><span class="p">[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()[:</span><span class="n">N</span><span class="p">]</span>

<span class="n">rows_rxn</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;smiles&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;react_prob&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;rationale_0&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;rationale_0_score&quot;</span><span class="p">:</span> <span class="p">[]}</span>

<span class="n">t0</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
<span class="k">for</span> <span class="n">smi</span> <span class="ow">in</span> <span class="n">sample_smis_rxn</span><span class="p">:</span>
    <span class="n">base_p</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">scoring_fn_rxn</span><span class="p">([</span><span class="n">smi</span><span class="p">])[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">rows_rxn</span><span class="p">[</span><span class="s2">&quot;smiles&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">smi</span><span class="p">)</span>
    <span class="n">rows_rxn</span><span class="p">[</span><span class="s2">&quot;react_prob&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">base_p</span><span class="p">)</span>

    <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">smi</span><span class="p">)</span>
    <span class="n">n</span> <span class="o">=</span> <span class="n">mol</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">()</span> <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="mi">20</span>
    <span class="n">min_atoms</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="mf">0.30</span> <span class="o">*</span> <span class="n">n</span><span class="p">))</span>
    <span class="n">max_atoms</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">min_atoms</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="mf">0.60</span> <span class="o">*</span> <span class="n">n</span><span class="p">))</span>

    <span class="n">rats</span> <span class="o">=</span> <span class="n">run_mcts_for_smiles</span><span class="p">(</span><span class="n">smi</span><span class="p">,</span> <span class="n">scoring_fn_rxn</span><span class="p">,</span> <span class="n">rollout</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">c_puct</span><span class="o">=</span><span class="mf">10.0</span><span class="p">,</span>
                               <span class="n">min_atoms</span><span class="o">=</span><span class="n">min_atoms</span><span class="p">,</span> <span class="n">max_atoms</span><span class="o">=</span><span class="n">max_atoms</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">rats</span><span class="p">:</span>
        <span class="n">ms</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">atoms</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">rats</span><span class="p">)</span>
        <span class="n">kept</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">rats</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">atoms</span><span class="p">)</span> <span class="o">==</span> <span class="n">ms</span> <span class="ow">and</span> <span class="n">x</span><span class="o">.</span><span class="n">smiles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">]</span>
        <span class="n">kept</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">P</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">r0</span> <span class="o">=</span> <span class="n">kept</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">smiles</span>
        <span class="n">r0s</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">kept</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">P</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">top1</span> <span class="o">=</span> <span class="n">fragment_top1</span><span class="p">(</span><span class="n">smi</span><span class="p">,</span> <span class="n">scoring_fn_rxn</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">top1</span><span class="p">:</span>
            <span class="n">r0</span><span class="p">,</span> <span class="n">r0s</span> <span class="o">=</span> <span class="n">top1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">r0</span><span class="p">,</span> <span class="n">r0s</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span>

    <span class="n">rows_rxn</span><span class="p">[</span><span class="s2">&quot;rationale_0&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r0</span><span class="p">)</span>
    <span class="n">rows_rxn</span><span class="p">[</span><span class="s2">&quot;rationale_0_score&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r0s</span><span class="p">)</span>

<span class="n">elapsed</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">t0</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reactivity: built table for </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">sample_smis_rxn</span><span class="p">)</span><span class="si">}</span><span class="s2"> molecules in </span><span class="si">{</span><span class="n">elapsed</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">s&quot;</span><span class="p">)</span>

<span class="n">reactivity_rationales</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">rows_rxn</span><span class="p">)</span>
<span class="n">reactivity_rationales</span>

<span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notna</span><span class="p">(</span><span class="n">reactivity_rationales</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;rationale_0&quot;</span><span class="p">]):</span>
    <span class="n">parent</span> <span class="o">=</span> <span class="n">reactivity_rationales</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;smiles&quot;</span><span class="p">]</span>
    <span class="n">rat</span>    <span class="o">=</span> <span class="n">reactivity_rationales</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;rationale_0&quot;</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Parent:&quot;</span><span class="p">,</span> <span class="n">parent</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Rationale 0:&quot;</span><span class="p">,</span> <span class="n">rat</span><span class="p">,</span> <span class="s2">&quot; score=&quot;</span><span class="p">,</span> <span class="n">reactivity_rationales</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;rationale_0_score&quot;</span><span class="p">])</span>
    <span class="n">visualize_rationale_on_parent</span><span class="p">(</span><span class="n">parent</span><span class="p">,</span> <span class="n">rat</span><span class="p">)</span>

</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="chemprop-cli-command-line-interface">
<h2><a class="toc-backref" href="#id8" role="doc-backlink">7. Chemprop CLI (Command-Line Interface)</a><a class="headerlink" href="#chemprop-cli-command-line-interface" title="Permalink to this heading">#</a></h2>
<p>Prepare a minimal CSV: <code class="docutils literal notranslate"><span class="pre">SMILES,Melting</span> <span class="pre">Point</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load data and write a small CSV</span>
<span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;https://raw.githubusercontent.com/zzhenglab/ai4chem/main/book/_data/C_H_oxidation_dataset.csv&quot;</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
<span class="n">reg_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">,</span> <span class="s2">&quot;Melting Point&quot;</span><span class="p">]</span>
<span class="n">df_reg</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">reg_cols</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">df_reg</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>SMILES</th>
      <th>Melting Point</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>c1ccc2c(c1)CCOC2</td>
      <td>65.8</td>
    </tr>
    <tr>
      <th>1</th>
      <td>c1ccc2c(c1)Cc1ccccc1-2</td>
      <td>90.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>c1ccc2c(c1)CCCC2</td>
      <td>69.4</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Save to disk for Chemprop CLI.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_reg</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;mp_data.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">len</span><span class="p">(</span><span class="n">df_reg</span><span class="p">),</span> <span class="n">df_reg</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(575,
                    SMILES  Melting Point
 0        c1ccc2c(c1)CCOC2           65.8
 1  c1ccc2c(c1)Cc1ccccc1-2           90.0)
</pre></div>
</div>
</div>
</div>
<p>Train a <strong>small</strong> model so it runs in class. We log common metrics.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># A short run. Increase epochs later if you have time/GPU.
!chemprop train \
  --data-path mp_data.csv \
  -t regression \
  -s SMILES \
  --target-columns &quot;Melting Point&quot; \
  -o mp_model \
  --num-replicates 1 \
  --epochs 15 \
  --save-smiles-splits \
  --metrics mae rmse r2 \
  --tracking-metric r2
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;chemprop&#39; is not recognized as an internal or external command,
operable program or batch file.
</pre></div>
</div>
</div>
</div>
<p>Make quick predictions on a few molecules.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>smiles_list = [
    &quot;CCO&quot;,              # ethanol
    &quot;c1ccccc1&quot;,         # benzene
    &quot;CC(=O)O&quot;,          # acetic acid
    &quot;CCN(CC)CC&quot;         # triethylamine
]
pd.DataFrame({&quot;SMILES&quot;: smiles_list}).to_csv(&quot;custom_smiles_reg.csv&quot;, index=False)

!chemprop predict \
  --test-path custom_smiles_reg.csv \
  --model-paths mp_model/replicate_0/model_0/best.pt \
  --preds-path mp_preds.csv

pd.read_csv(&quot;mp_preds.csv&quot;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;chemprop&#39; is not recognized as an internal or external command,
operable program or batch file.
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>SMILES</th>
      <th>MP</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CCO</td>
      <td>112.653930</td>
    </tr>
    <tr>
      <th>1</th>
      <td>c1ccccc1</td>
      <td>113.069084</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CC(=O)O</td>
      <td>112.723390</td>
    </tr>
    <tr>
      <th>3</th>
      <td>O=C(O)C(O)C(O)C(O)CO</td>
      <td>124.462326</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<section id="reactivity-classification-ch-oxidation-dataset">
<h3>Reactivity classification (C–H oxidation dataset)<a class="headerlink" href="#reactivity-classification-ch-oxidation-dataset" title="Permalink to this heading">#</a></h3>
<p>We use the <code class="docutils literal notranslate"><span class="pre">Reactivity</span></code> column and convert it to <strong>binary</strong> 0/1.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
<span class="n">df</span><span class="p">[</span><span class="s2">&quot;Reactivity_bin&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;Reactivity&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">({</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span> <span class="mi">0</span><span class="p">})</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<span class="n">df</span><span class="p">[[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">,</span><span class="s2">&quot;Reactivity&quot;</span><span class="p">,</span><span class="s2">&quot;Reactivity_bin&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>SMILES</th>
      <th>Reactivity</th>
      <th>Reactivity_bin</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>c1ccc2c(c1)CCOC2</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>c1ccc2c(c1)Cc1ccccc1-2</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>c1ccc2c(c1)CCCC2</td>
      <td>1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Write a minimal file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="p">[[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">,</span> <span class="s2">&quot;Reactivity_bin&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;reactivity_data_bin.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># Optional: sanity check the class balance</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;Reactivity&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">to_dict</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;Reactivity_bin&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">to_dict</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{-1: 311, 1: 264}
{0: 311, 1: 264}
</pre></div>
</div>
</div>
</div>
<p>Train a short classification model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>!chemprop train \
  --data-path reactivity_data_bin.csv \
  -t classification \
  -s SMILES \
  --target-columns Reactivity_bin \
  -o reactivity_model \
  --num-replicates 1 \
  --epochs 15 \
  --class-balance \
  --metrics roc prc accuracy \
  --tracking-metric roc
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;chemprop&#39; is not recognized as an internal or external command,
operable program or batch file.
</pre></div>
</div>
</div>
</div>
<p>Predict on new SMILES.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>smiles_list = [
    &quot;CCO&quot;,
    &quot;c1ccccc1C(F)&quot;,
    &quot;C1=C([C@@H]2C[C@H](C1)C2(C)C)&quot;,
    &quot;C1=CC=CC=C1C=O&quot;,
    &quot;CCN(CC)CC&quot;,
    &quot;c1cccc(C=CC)c1&quot;
]
pd.DataFrame({&quot;SMILES&quot;: smiles_list}).to_csv(&quot;custom_smiles.csv&quot;, index=False)

!chemprop predict \
  --test-path custom_smiles.csv \
  --model-paths reactivity_model/replicate_0/model_0/best.pt \
  --preds-path custom_preds.csv

pd.read_csv(&quot;custom_preds.csv&quot;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;chemprop&#39; is not recognized as an internal or external command,
operable program or batch file.
</pre></div>
</div>
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">FileNotFoundError</span><span class="g g-Whitespace">                         </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">16</span><span class="p">],</span> <span class="n">line</span> <span class="mi">13</span>
<span class="g g-Whitespace">      </span><span class="mi">9</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;SMILES&quot;</span><span class="p">:</span> <span class="n">smiles_list</span><span class="p">})</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;custom_smiles.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">11</span> <span class="n">get_ipython</span><span class="p">()</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="s1">&#39;chemprop predict    --test-path custom_smiles.csv    --model-paths reactivity_model/replicate_0/model_0/best.pt    --preds-path custom_preds.csv&#39;</span><span class="p">)</span>
<span class="ne">---&gt; </span><span class="mi">13</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;custom_preds.csv&quot;</span><span class="p">)</span>

<span class="nn">File c:\users\52377\appdata\local\programs\python\python38\lib\site-packages\pandas\io\parsers\readers.py:912,</span> in <span class="ni">read_csv</span><span class="nt">(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)</span>
<span class="g g-Whitespace">    </span><span class="mi">899</span> <span class="n">kwds_defaults</span> <span class="o">=</span> <span class="n">_refine_defaults_read</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">900</span>     <span class="n">dialect</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">901</span>     <span class="n">delimiter</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">908</span>     <span class="n">dtype_backend</span><span class="o">=</span><span class="n">dtype_backend</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">909</span> <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">910</span> <span class="n">kwds</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwds_defaults</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">912</span> <span class="k">return</span> <span class="n">_read</span><span class="p">(</span><span class="n">filepath_or_buffer</span><span class="p">,</span> <span class="n">kwds</span><span class="p">)</span>

<span class="nn">File c:\users\52377\appdata\local\programs\python\python38\lib\site-packages\pandas\io\parsers\readers.py:577,</span> in <span class="ni">_read</span><span class="nt">(filepath_or_buffer, kwds)</span>
<span class="g g-Whitespace">    </span><span class="mi">574</span> <span class="n">_validate_names</span><span class="p">(</span><span class="n">kwds</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;names&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">))</span>
<span class="g g-Whitespace">    </span><span class="mi">576</span> <span class="c1"># Create the parser.</span>
<span class="ne">--&gt; </span><span class="mi">577</span> <span class="n">parser</span> <span class="o">=</span> <span class="n">TextFileReader</span><span class="p">(</span><span class="n">filepath_or_buffer</span><span class="p">,</span> <span class="o">**</span><span class="n">kwds</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">579</span> <span class="k">if</span> <span class="n">chunksize</span> <span class="ow">or</span> <span class="n">iterator</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">580</span>     <span class="k">return</span> <span class="n">parser</span>

<span class="nn">File c:\users\52377\appdata\local\programs\python\python38\lib\site-packages\pandas\io\parsers\readers.py:1407,</span> in <span class="ni">TextFileReader.__init__</span><span class="nt">(self, f, engine, **kwds)</span>
<span class="g g-Whitespace">   </span><span class="mi">1404</span>     <span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="p">[</span><span class="s2">&quot;has_index_names&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwds</span><span class="p">[</span><span class="s2">&quot;has_index_names&quot;</span><span class="p">]</span>
<span class="g g-Whitespace">   </span><span class="mi">1406</span> <span class="bp">self</span><span class="o">.</span><span class="n">handles</span><span class="p">:</span> <span class="n">IOHandles</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
<span class="ne">-&gt; </span><span class="mi">1407</span> <span class="bp">self</span><span class="o">.</span><span class="n">_engine</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_make_engine</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">engine</span><span class="p">)</span>

<span class="nn">File c:\users\52377\appdata\local\programs\python\python38\lib\site-packages\pandas\io\parsers\readers.py:1661,</span> in <span class="ni">TextFileReader._make_engine</span><span class="nt">(self, f, engine)</span>
<span class="g g-Whitespace">   </span><span class="mi">1659</span>     <span class="k">if</span> <span class="s2">&quot;b&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mode</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1660</span>         <span class="n">mode</span> <span class="o">+=</span> <span class="s2">&quot;b&quot;</span>
<span class="ne">-&gt; </span><span class="mi">1661</span> <span class="bp">self</span><span class="o">.</span><span class="n">handles</span> <span class="o">=</span> <span class="n">get_handle</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1662</span>     <span class="n">f</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1663</span>     <span class="n">mode</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1664</span>     <span class="n">encoding</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;encoding&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
<span class="g g-Whitespace">   </span><span class="mi">1665</span>     <span class="n">compression</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;compression&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
<span class="g g-Whitespace">   </span><span class="mi">1666</span>     <span class="n">memory_map</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;memory_map&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">),</span>
<span class="g g-Whitespace">   </span><span class="mi">1667</span>     <span class="n">is_text</span><span class="o">=</span><span class="n">is_text</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1668</span>     <span class="n">errors</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;encoding_errors&quot;</span><span class="p">,</span> <span class="s2">&quot;strict&quot;</span><span class="p">),</span>
<span class="g g-Whitespace">   </span><span class="mi">1669</span>     <span class="n">storage_options</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;storage_options&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
<span class="g g-Whitespace">   </span><span class="mi">1670</span> <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1671</span> <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">handles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
<span class="g g-Whitespace">   </span><span class="mi">1672</span> <span class="n">f</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">handles</span><span class="o">.</span><span class="n">handle</span>

<span class="nn">File c:\users\52377\appdata\local\programs\python\python38\lib\site-packages\pandas\io\common.py:859,</span> in <span class="ni">get_handle</span><span class="nt">(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)</span>
<span class="g g-Whitespace">    </span><span class="mi">854</span> <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
<span class="g g-Whitespace">    </span><span class="mi">855</span>     <span class="c1"># Check whether the filename is to be opened in binary mode.</span>
<span class="g g-Whitespace">    </span><span class="mi">856</span>     <span class="c1"># Binary mode does not support &#39;encoding&#39; and &#39;newline&#39;.</span>
<span class="g g-Whitespace">    </span><span class="mi">857</span>     <span class="k">if</span> <span class="n">ioargs</span><span class="o">.</span><span class="n">encoding</span> <span class="ow">and</span> <span class="s2">&quot;b&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ioargs</span><span class="o">.</span><span class="n">mode</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">858</span>         <span class="c1"># Encoding</span>
<span class="ne">--&gt; </span><span class="mi">859</span>         <span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">860</span>             <span class="n">handle</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">861</span>             <span class="n">ioargs</span><span class="o">.</span><span class="n">mode</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">862</span>             <span class="n">encoding</span><span class="o">=</span><span class="n">ioargs</span><span class="o">.</span><span class="n">encoding</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">863</span>             <span class="n">errors</span><span class="o">=</span><span class="n">errors</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">864</span>             <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">865</span>         <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">866</span>     <span class="k">else</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">867</span>         <span class="c1"># Binary mode</span>
<span class="g g-Whitespace">    </span><span class="mi">868</span>         <span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="n">ioargs</span><span class="o">.</span><span class="n">mode</span><span class="p">)</span>

<span class="ne">FileNotFoundError</span>: [Errno 2] No such file or directory: &#39;custom_preds.csv&#39;
</pre></div>
</div>
</div>
</div>
<div class="admonition-tips admonition">
<p class="admonition-title">Tips</p>
<ul class="simple">
<li><p>Increase <code class="docutils literal notranslate"><span class="pre">--num-replicates</span></code> to 3 and <code class="docutils literal notranslate"><span class="pre">--epochs</span></code> to 50-100 for stronger baselines.</p></li>
<li><p>For class imbalance, keep <code class="docutils literal notranslate"><span class="pre">--class-balance</span></code>.</p></li>
<li><p>Use <code class="docutils literal notranslate"><span class="pre">--save-smiles-splits</span></code> to capture exact train/val/test molecules for reproducibility.</p></li>
</ul>
</div>
<div class="admonition-exercises-7-x admonition">
<p class="admonition-title">⏰ Exercises 7.x</p>
<ol class="arabic simple">
<li><p>Add <code class="docutils literal notranslate"><span class="pre">--ensemble-size</span> <span class="pre">5</span></code> during prediction by passing multiple <code class="docutils literal notranslate"><span class="pre">--model-paths</span></code> if you trained replicates. Compare ROC.</p></li>
<li><p>Change tracking metric to <code class="docutils literal notranslate"><span class="pre">prc</span></code> and rerun. Does validation selection change.</p></li>
<li><p>For melting point, add <code class="docutils literal notranslate"><span class="pre">--ffn-hidden-size</span> <span class="pre">800</span></code> to increase the head capacity and try 30 epochs.</p></li>
</ol>
</div>
</section>
</section>
<hr class="docutils" />
<section id="glossary-references-in-class-activity-with-solutions">
<h2><a class="toc-backref" href="#id9" role="doc-backlink">10. Glossary • References • In-class activity with solutions</a><a class="headerlink" href="#glossary-references-in-class-activity-with-solutions" title="Permalink to this heading">#</a></h2>
<section id="glossary">
<h3>10.1 Glossary<a class="headerlink" href="#glossary" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><strong>MPNN</strong>: Message Passing Neural Network. Learns node and edge embeddings by exchanging information along bonds.</p></li>
<li><p><strong>Aggregation</strong>: Pools node embeddings to a molecule vector. Common choices are mean and sum.</p></li>
<li><p><strong>FFN</strong>: Feed-forward network head used for regression or classification.</p></li>
<li><p><strong>AUROC</strong>: Area under ROC. Measures ranking quality of a binary classifier.</p></li>
<li><p><strong>Selectivity (atom-level)</strong>: Probability that a given atom is the reaction site under a given transformation.</p></li>
<li><p><strong>CGR</strong>: Condensed Graph of Reaction. A featurization for reaction tasks.</p></li>
<li><p><strong>SHAP</strong>: Shapley values for local explanation of a prediction.</p></li>
</ul>
</section>
<section id="references">
<h3>10.2 References<a class="headerlink" href="#references" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Li, S.-C., Wu, H., Menon, A., Spiekermann, K. A., Li, Y.-P., Green, W. H. When Do Quantum Mechanical Descriptors Help Graph Neural Networks to Predict Chemical Properties? <strong>JACS</strong> 2024, 146, 23103–23120. <a class="reference external" href="https://doi.org/10.1021/jacs.4c04670">https://doi.org/10.1021/jacs.4c04670</a><br />
SHAP implementation for Chemprop v2 is based on the approach described by the authors.</p></li>
<li><p>Chemprop v2 documentation and examples: training, classification, multicomponent, reaction and atom/bond prediction.</p></li>
</ul>
</section>
<section id="in-class-activity">
<h3>10.3 In-class activity<a class="headerlink" href="#in-class-activity" title="Permalink to this heading">#</a></h3>
<p>Answer all five questions. Most are small edits of code we wrote today.</p>
<p><strong>Q1. Solubility log-transform</strong><br />
Create <code class="docutils literal notranslate"><span class="pre">y_log</span> <span class="pre">=</span> <span class="pre">log10(Solubility_mol_per_L</span> <span class="pre">+</span> <span class="pre">1e-6)</span></code> and train a regression MPNN. Compare RMSE to the non-log model.</p>
<p><strong>Q2. pKa vs Melting multi-compare</strong><br />
Train two separate models as done above and report test MAE for each. Which one is easier to predict? Add a parity plot for both.</p>
<p><strong>Q3. Toxicity thresholding</strong><br />
Change the classification threshold from 0.5 to the value that maximizes F1 on the validation set. Report the new test accuracy and AUROC.</p>
<p><strong>Q4. Reactivity calibration</strong><br />
For the reactivity model, plot a reliability curve (predicted prob vs empirical rate binned into deciles). Is the model over or under confident?</p>
<p><strong>Q5. Atom-level selectivity</strong><br />
Take five molecules from the test set with nonempty <code class="docutils literal notranslate"><span class="pre">site_list</span></code>. For each, show the top 2 atoms by predicted probability and compare with ground truth indices.</p>
</section>
<hr class="docutils" />
<section id="solutions">
<h3>10.4 Solutions<a class="headerlink" href="#solutions" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Log-solubility</span>
<span class="n">df_log</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">,</span><span class="s2">&quot;Solubility_mol_per_L&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">df_log</span><span class="p">[</span><span class="s2">&quot;y_log&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">df_log</span><span class="p">[</span><span class="s2">&quot;Solubility_mol_per_L&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="mf">1e-6</span><span class="p">)</span>

<span class="n">smis</span> <span class="o">=</span> <span class="n">df_log</span><span class="p">[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">ys</span>   <span class="o">=</span> <span class="n">df_log</span><span class="p">[</span><span class="s2">&quot;y_log&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
<span class="n">dps</span> <span class="o">=</span> <span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">MoleculeDatapoint</span><span class="o">.</span><span class="n">from_smi</span><span class="p">(</span><span class="n">s</span><span class="p">,</span><span class="n">y</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span><span class="p">,</span><span class="n">y</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">smis</span><span class="p">,</span> <span class="n">ys</span><span class="p">)]</span>
<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="n">dp</span><span class="o">.</span><span class="n">mol</span> <span class="k">for</span> <span class="n">dp</span> <span class="ow">in</span> <span class="n">dps</span><span class="p">]</span>
<span class="n">tr_idx</span><span class="p">,</span> <span class="n">va_idx</span><span class="p">,</span> <span class="n">te_idx</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">make_split_indices</span><span class="p">(</span><span class="n">mols</span><span class="p">,</span> <span class="s2">&quot;random&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mf">0.8</span><span class="p">,</span><span class="mf">0.1</span><span class="p">,</span><span class="mf">0.1</span><span class="p">))</span>
<span class="n">tr</span><span class="p">,</span> <span class="n">va</span><span class="p">,</span> <span class="n">te</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">split_data_by_indices</span><span class="p">(</span><span class="n">dps</span><span class="p">,</span> <span class="n">tr_idx</span><span class="p">,</span> <span class="n">va_idx</span><span class="p">,</span> <span class="n">te_idx</span><span class="p">)</span>

<span class="n">feat</span> <span class="o">=</span> <span class="n">featurizers</span><span class="o">.</span><span class="n">SimpleMoleculeMolGraphFeaturizer</span><span class="p">()</span>
<span class="n">tr_set</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MoleculeDataset</span><span class="p">(</span><span class="n">tr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>
<span class="n">sc</span> <span class="o">=</span> <span class="n">tr_set</span><span class="o">.</span><span class="n">normalize_targets</span><span class="p">()</span>
<span class="n">va_set</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MoleculeDataset</span><span class="p">(</span><span class="n">va</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">);</span> <span class="n">va_set</span><span class="o">.</span><span class="n">normalize_targets</span><span class="p">(</span><span class="n">sc</span><span class="p">)</span>
<span class="n">te_set</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MoleculeDataset</span><span class="p">(</span><span class="n">te</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">feat</span><span class="p">)</span>

<span class="n">mp</span>  <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">BondMessagePassing</span><span class="p">()</span>
<span class="n">agg</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">MeanAggregation</span><span class="p">()</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">RegressionFFN</span><span class="p">(</span><span class="n">output_transform</span><span class="o">=</span><span class="n">nn</span><span class="o">.</span><span class="n">UnscaleTransform</span><span class="o">.</span><span class="n">from_standard_scaler</span><span class="p">(</span><span class="n">sc</span><span class="p">))</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">models</span><span class="o">.</span><span class="n">MPNN</span><span class="p">(</span><span class="n">mp</span><span class="p">,</span> <span class="n">agg</span><span class="p">,</span> <span class="n">out</span><span class="p">,</span> <span class="n">batch_norm</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metrics</span><span class="o">=</span><span class="p">[</span><span class="n">nn</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">RMSE</span><span class="p">(),</span> <span class="n">nn</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">MAE</span><span class="p">()])</span>

<span class="n">tr_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">tr_set</span><span class="p">)</span>
<span class="n">va_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">va_set</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">te_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">te_set</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">trainer_log</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">Trainer</span><span class="p">(</span><span class="n">logger</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">enable_checkpointing</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">accelerator</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span>
                         <span class="n">devices</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_epochs</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">trainer_log</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">tr_loader</span><span class="p">,</span> <span class="n">va_loader</span><span class="p">)</span>
<span class="n">trainer_log</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">te_loader</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># pKa</span>
<span class="n">pka_tr</span><span class="p">,</span> <span class="n">pka_va</span><span class="p">,</span> <span class="n">pka_te</span><span class="p">,</span> <span class="n">pka_un</span> <span class="o">=</span> <span class="n">build_regression_loaders</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">&quot;pKa&quot;</span><span class="p">)</span>
<span class="n">mpnn_pka2</span><span class="p">,</span> <span class="n">trainer_pka2</span><span class="p">,</span> <span class="n">pka_te_loader2</span><span class="p">,</span> <span class="n">pka_te_set2</span> <span class="o">=</span> <span class="n">train_regression_mpn</span><span class="p">(</span>
    <span class="n">pka_tr</span><span class="p">,</span> <span class="n">pka_va</span><span class="p">,</span> <span class="n">pka_te</span><span class="p">,</span> <span class="n">pka_un</span><span class="p">,</span> <span class="n">epochs</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">tag</span><span class="o">=</span><span class="s2">&quot;pka_q2&quot;</span>
<span class="p">)</span>

<span class="c1"># Melting</span>
<span class="n">mp_tr</span><span class="p">,</span> <span class="n">mp_va</span><span class="p">,</span> <span class="n">mp_te</span><span class="p">,</span> <span class="n">mp_un</span> <span class="o">=</span> <span class="n">build_regression_loaders</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">&quot;Melting Point&quot;</span><span class="p">)</span>
<span class="n">mpnn_mp2</span><span class="p">,</span> <span class="n">trainer_mp2</span><span class="p">,</span> <span class="n">mp_te_loader2</span><span class="p">,</span> <span class="n">mp_te_set2</span> <span class="o">=</span> <span class="n">train_regression_mpn</span><span class="p">(</span>
    <span class="n">mp_tr</span><span class="p">,</span> <span class="n">mp_va</span><span class="p">,</span> <span class="n">mp_te</span><span class="p">,</span> <span class="n">mp_un</span><span class="p">,</span> <span class="n">epochs</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">tag</span><span class="o">=</span><span class="s2">&quot;mp_q2&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">f1_score</span>

<span class="c1"># Get validation probabilities and find best threshold</span>
<span class="n">va_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">tox_va</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">inference_mode</span><span class="p">():</span>
    <span class="n">va_probs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">trainer_tox</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">mpnn_tox</span><span class="p">,</span> <span class="n">va_loader</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="n">y_val</span> <span class="o">=</span> <span class="n">tox_va</span><span class="o">.</span><span class="n">Y</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

<span class="n">ths</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">,</span> <span class="mi">41</span><span class="p">)</span>
<span class="n">f1s</span> <span class="o">=</span> <span class="p">[</span><span class="n">f1_score</span><span class="p">(</span><span class="n">y_val</span><span class="p">,</span> <span class="p">(</span><span class="n">va_probs</span><span class="o">&gt;=</span><span class="n">t</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">))</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">ths</span><span class="p">]</span>
<span class="n">best_t</span> <span class="o">=</span> <span class="n">ths</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">f1s</span><span class="p">))]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Best threshold by F1:&quot;</span><span class="p">,</span> <span class="n">best_t</span><span class="p">)</span>

<span class="c1"># Evaluate on test</span>
<span class="n">te_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">tox_te</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">inference_mode</span><span class="p">():</span>
    <span class="n">te_probs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">trainer_tox</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">mpnn_tox</span><span class="p">,</span> <span class="n">te_loader</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="n">y_test</span> <span class="o">=</span> <span class="n">tox_te</span><span class="o">.</span><span class="n">Y</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

<span class="n">acc</span> <span class="o">=</span> <span class="n">accuracy_score</span><span class="p">(</span><span class="n">y_test</span><span class="p">,</span> <span class="p">(</span><span class="n">te_probs</span><span class="o">&gt;=</span><span class="n">best_t</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">))</span>
<span class="n">auc</span> <span class="o">=</span> <span class="n">roc_auc_score</span><span class="p">(</span><span class="n">y_test</span><span class="p">,</span> <span class="n">te_probs</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;New test Accuracy: </span><span class="si">{</span><span class="n">acc</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">  AUC: </span><span class="si">{</span><span class="n">auc</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Reliability curve for reactivity</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.calibration</span><span class="w"> </span><span class="kn">import</span> <span class="n">calibration_curve</span>

<span class="n">te_loader</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">rxn_te</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">inference_mode</span><span class="p">():</span>
    <span class="n">re_probs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">trainer_rxn</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">mpnn_rxn</span><span class="p">,</span> <span class="n">te_loader</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">rxn_te</span><span class="o">.</span><span class="n">Y</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

<span class="n">frac_pos</span><span class="p">,</span> <span class="n">mean_pred</span> <span class="o">=</span> <span class="n">calibration_curve</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">re_probs</span><span class="p">,</span> <span class="n">n_bins</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">strategy</span><span class="o">=</span><span class="s2">&quot;quantile&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">mean_pred</span><span class="p">,</span> <span class="n">frac_pos</span><span class="p">,</span> <span class="s2">&quot;o-&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span><span class="s2">&quot;k--&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Mean predicted probability&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Fraction positive&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Reliability curve: Reactivity&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Pick five with nonempty site_list</span>
<span class="n">cand</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;site_list&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="nb">len</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>

<span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">smi</span><span class="p">,</span> <span class="n">sites</span> <span class="ow">in</span> <span class="n">cand</span><span class="p">[[</span><span class="s2">&quot;SMILES&quot;</span><span class="p">,</span><span class="s2">&quot;site_list&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">itertuples</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="n">dp</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MolAtomBondDatapoint</span><span class="o">.</span><span class="n">from_smi</span><span class="p">(</span><span class="n">smi</span><span class="p">,</span> <span class="n">reorder_atoms</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ds</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">MolAtomBondDataset</span><span class="p">([</span><span class="n">dp</span><span class="p">],</span> <span class="n">featurizer</span><span class="o">=</span><span class="n">featurizers</span><span class="o">.</span><span class="n">SimpleMoleculeMolGraphFeaturizer</span><span class="p">())</span>
    <span class="n">dl</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">build_dataloader</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">inference_mode</span><span class="p">():</span>
        <span class="n">batch</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">dl</span><span class="p">))</span>
        <span class="n">out</span> <span class="o">=</span> <span class="n">model_sel</span><span class="p">(</span><span class="o">*</span><span class="n">batch</span><span class="p">[:</span><span class="mi">1</span><span class="p">])</span> <span class="c1"># mg only</span>
        <span class="n">probs</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">sigmoid</span><span class="p">(</span><span class="n">out</span><span class="p">[</span><span class="s2">&quot;atom&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
    <span class="n">top2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">probs</span><span class="p">)[</span><span class="o">-</span><span class="mi">2</span><span class="p">:]</span> <span class="o">+</span> <span class="mi">1</span>  <span class="c1"># convert to 1-based for comparison</span>
    <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s2">&quot;SMILES&quot;</span><span class="p">:</span> <span class="n">smi</span><span class="p">,</span> <span class="s2">&quot;true_sites&quot;</span><span class="p">:</span> <span class="n">sites</span><span class="p">,</span> <span class="s2">&quot;top2_pred&quot;</span><span class="p">:</span> <span class="n">top2</span><span class="o">.</span><span class="n">tolist</span><span class="p">()})</span>

<span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-goals">Learning goals</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#directed-message-passing-neural-network-d-mpnn">1. Directed message-passing neural network (D-MPNN)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-preparation">2. Data preparation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#property-prediction-regression">3. Property prediction (regression)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pick-blocks">3.1 Pick blocks</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#build-model-and-trainer">3.2 Build model and trainer</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#train">3.3 Train</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#test-and-parity-plot">3.4 Test and parity plot</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#property-prediction-classification">4. Property prediction (classification)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#build-classification-dataset">4.1 Build classification dataset</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#model-and-training">4.2 Model and training</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#roc-curve">4.3 ROC curve</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reactivity-and-selectivity">5. Reactivity and selectivity</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reactivity-classifier">5.1 Reactivity classifier</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#atom-level-selectivity-build-labels-per-atom">5.2 Atom-level selectivity: build labels per atom</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#rationale">6. Rationale</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#solubility-prediction-rationale">6.1 Solubility prediction rationale</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reactivity-prediction-rationale">6.2 Reactivity prediction rationale</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#chemprop-cli-command-line-interface">7. Chemprop CLI (Command-Line Interface)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reactivity-classification-ch-oxidation-dataset">Reactivity classification (C–H oxidation dataset)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#glossary-references-in-class-activity-with-solutions">10. Glossary • References • In-class activity with solutions</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#glossary">10.1 Glossary</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#references">10.2 References</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#in-class-activity">10.3 In-class activity</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#solutions">10.4 Solutions</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The Jupyter Book Community
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>